// auth/dist/index.mjs
#!/usr/bin/env node
import { a as getPackageInfo, c as spawnCommand, d as generateDrizzleSchema, i as findMonorepoRoot, l as tryCatch, n as generateSchema, o as hasDependency, r as generatePrismaSchema, s as generateSecretHash$1 } from "./generators-DNY9D4Si.mjs";
import { Command } from "commander";
import * as fs$2 from "node:fs";
import fs, { existsSync, readFileSync, readdirSync } from "node:fs";
import fs$1 from "node:fs/promises";
import * as path$1 from "node:path";
import path, { join } from "node:path";
import { createTelemetry, getTelemetryAuthConfig } from "@better-auth/telemetry";
import { getAdapter } from "better-auth/db/adapter";
import chalk from "chalk";
import prompts from "prompts";
import yoctoSpinner from "yocto-spinner";
import * as z$1 from "zod/v4";
import { format } from "prettier";
import { getMigrations } from "better-auth/db/migration";
import { exec, execSync } from "node:child_process";
import Crypto from "node:crypto";
import babelPresetReact from "@babel/preset-react";
import babelPresetTypeScript from "@babel/preset-typescript";
import { BetterAuthError } from "@better-auth/core/error";
import { loadConfig } from "c12";
import * as os$1 from "node:os";
import os from "node:os";
import open from "open";
import z from "zod";
import { env } from "@better-auth/core/env";
import { log } from "@clack/prompts";
import { base64 } from "@better-auth/utils/base64";
import "dotenv/config";

//#region src/utils/add-cloudflare-modules.ts
const createModule = () => {
	return `data:text/javascript;charset=utf-8,${encodeURIComponent(`
const createStub = (label) => {
  const handler = {
    get(_, prop) {
      if (prop === "toString") return () => label;
      if (prop === "valueOf") return () => label;
      if (prop === Symbol.toPrimitive) return () => label;
      if (prop === Symbol.toStringTag) return "Object";
      if (prop === "then") return undefined;
      return createStub(label + "." + String(prop));
    },
    apply(_, __, args) {
      return createStub(label + "()")
    },
    construct() {
      return createStub(label + "#instance");
    },
  };
  const fn = () => createStub(label + "()");
  return new Proxy(fn, handler);
};

class WorkerEntrypoint {
  constructor(ctx, env) {
    this.ctx = ctx;
    this.env = env;
  }
}

class DurableObject {
  constructor(state, env) {
    this.state = state;
    this.env = env;
  }
}

class RpcTarget {
  constructor(value) {
    this.value = value;
  }
}

const RpcStub = RpcTarget;

const env = createStub("env");
const caches = createStub("caches");
const scheduler = createStub("scheduler");
const executionCtx = createStub("executionCtx");

export { DurableObject, RpcStub, RpcTarget, WorkerEntrypoint, caches, env, executionCtx, scheduler };

const defaultExport = {
  DurableObject,
  RpcStub,
  RpcTarget,
  WorkerEntrypoint,
  caches,
  env,
  executionCtx,
  scheduler,
};

export default defaultExport;
// jiti dirty hack: .unknown
`)}`;
};
const CLOUDFLARE_STUB_MODULE = createModule();
function addCloudflareModules(aliases, _cwd) {
	if (!aliases["cloudflare:workers"]) aliases["cloudflare:workers"] = CLOUDFLARE_STUB_MODULE;
	if (!aliases["cloudflare:test"]) aliases["cloudflare:test"] = CLOUDFLARE_STUB_MODULE;
}

//#endregion
//#region src/utils/add-svelte-kit-env-modules.ts
/**
* Adds SvelteKit environment modules and path aliases
* @param aliases - The aliases object to populate
* @param cwd - Current working directory (optional, defaults to process.cwd())
*/
function addSvelteKitEnvModules(aliases, cwd) {
	const workingDir = cwd || process.cwd();
	aliases["$env/dynamic/private"] = createDataUriModule(createDynamicEnvModule());
	aliases["$env/dynamic/public"] = createDataUriModule(createDynamicEnvModule());
	aliases["$env/static/private"] = createDataUriModule(createStaticEnvModule(filterPrivateEnv("PUBLIC_", "")));
	aliases["$env/static/public"] = createDataUriModule(createStaticEnvModule(filterPublicEnv("PUBLIC_", "")));
	const svelteKitAliases = getSvelteKitPathAliases(workingDir);
	Object.assign(aliases, svelteKitAliases);
}
function getSvelteKitPathAliases(cwd) {
	const aliases = {};
	const packageJsonPath = path.join(cwd, "package.json");
	const svelteConfigPath = path.join(cwd, "svelte.config.js");
	const svelteConfigTsPath = path.join(cwd, "svelte.config.ts");
	let isSvelteKitProject = false;
	if (fs.existsSync(packageJsonPath)) try {
		const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
		isSvelteKitProject = !!{
			...packageJson.dependencies,
			...packageJson.devDependencies
		}["@sveltejs/kit"];
	} catch {}
	if (!isSvelteKitProject) isSvelteKitProject = fs.existsSync(svelteConfigPath) || fs.existsSync(svelteConfigTsPath);
	if (!isSvelteKitProject) return aliases;
	const libPaths = [path.join(cwd, "src", "lib"), path.join(cwd, "lib")];
	for (const libPath of libPaths) if (fs.existsSync(libPath)) {
		aliases["$lib"] = libPath;
		for (const subPath of [
			"server",
			"utils",
			"components",
			"stores"
		]) {
			const subDir = path.join(libPath, subPath);
			if (fs.existsSync(subDir)) aliases[`$lib/${subPath}`] = subDir;
		}
		break;
	}
	aliases["$app/server"] = createDataUriModule(createAppServerModule());
	const customAliases = getSvelteConfigAliases(cwd);
	Object.assign(aliases, customAliases);
	return aliases;
}
function getSvelteConfigAliases(cwd) {
	const aliases = {};
	const configPaths = [path.join(cwd, "svelte.config.js"), path.join(cwd, "svelte.config.ts")];
	for (const configPath of configPaths) if (fs.existsSync(configPath)) {
		try {
			const aliasMatch = fs.readFileSync(configPath, "utf-8").match(/alias\s*:\s*\{([^}]+)\}/);
			if (aliasMatch && aliasMatch[1]) {
				const aliasMatches = aliasMatch[1].matchAll(/['"`](\$[^'"`]+)['"`]\s*:\s*['"`]([^'"`]+)['"`]/g);
				for (const match of aliasMatches) {
					const [, alias, target] = match;
					if (alias && target) {
						aliases[alias + "/*"] = path.resolve(cwd, target) + "/*";
						aliases[alias] = path.resolve(cwd, target);
					}
				}
			}
		} catch {}
		break;
	}
	return aliases;
}
function createAppServerModule() {
	return `
// $app/server stub for CLI compatibility
export default {};
// jiti dirty hack: .unknown
`;
}
function createDataUriModule(module) {
	return `data:text/javascript;charset=utf-8,${encodeURIComponent(module)}`;
}
function createStaticEnvModule(env) {
	return `
  ${Object.keys(env).filter((k) => validIdentifier.test(k) && !reserved.has(k)).map((k) => `export const ${k} = ${JSON.stringify(env[k])};`).join("\n")}
  // jiti dirty hack: .unknown
  `;
}
function createDynamicEnvModule() {
	return `
  export const env = process.env;
  // jiti dirty hack: .unknown
  `;
}
function filterPrivateEnv(publicPrefix, privatePrefix) {
	return Object.fromEntries(Object.entries(process.env).filter(([k]) => k.startsWith(privatePrefix) && (publicPrefix === "" || !k.startsWith(publicPrefix))));
}
function filterPublicEnv(publicPrefix, privatePrefix) {
	return Object.fromEntries(Object.entries(process.env).filter(([k]) => k.startsWith(publicPrefix) && (privatePrefix === "" || !k.startsWith(privatePrefix))));
}
const validIdentifier = /^[a-zA-Z_$][a-zA-Z0-9_$]*$/;
const reserved = new Set([
	"do",
	"if",
	"in",
	"for",
	"let",
	"new",
	"try",
	"var",
	"case",
	"else",
	"enum",
	"eval",
	"null",
	"this",
	"true",
	"void",
	"with",
	"await",
	"break",
	"catch",
	"class",
	"const",
	"false",
	"super",
	"throw",
	"while",
	"yield",
	"delete",
	"export",
	"import",
	"public",
	"return",
	"static",
	"switch",
	"typeof",
	"default",
	"extends",
	"finally",
	"package",
	"private",
	"continue",
	"debugger",
	"function",
	"arguments",
	"interface",
	"protected",
	"implements",
	"instanceof"
]);

//#endregion
//#region src/utils/get-tsconfig-info.ts
function stripJsonComments(jsonString) {
	return jsonString.replace(/\\"|"(?:\\"|[^"])*"|(\/\/.*|\/\*[\s\S]*?\*\/)/g, (m, g) => g ? "" : m).replace(/,(?=\s*[}\]])/g, "");
}
function getTsconfigInfo(cwd, flatPath) {
	let tsConfigPath;
	if (flatPath) tsConfigPath = flatPath;
	else tsConfigPath = cwd ? path.join(cwd, "tsconfig.json") : path.join("tsconfig.json");
	try {
		const text = fs.readFileSync(tsConfigPath, "utf-8");
		return JSON.parse(stripJsonComments(text));
	} catch (error) {
		throw error;
	}
}

//#endregion
//#region src/utils/get-config.ts
let possiblePaths$1 = [
	"auth.ts",
	"auth.tsx",
	"auth.js",
	"auth.jsx",
	"auth.server.js",
	"auth.server.ts",
	"auth/index.ts",
	"auth/index.tsx",
	"auth/index.js",
	"auth/index.jsx",
	"auth/index.server.js",
	"auth/index.server.ts"
];
possiblePaths$1 = [
	...possiblePaths$1,
	...possiblePaths$1.map((it) => `lib/server/${it}`),
	...possiblePaths$1.map((it) => `server/auth/${it}`),
	...possiblePaths$1.map((it) => `server/${it}`),
	...possiblePaths$1.map((it) => `auth/${it}`),
	...possiblePaths$1.map((it) => `lib/${it}`),
	...possiblePaths$1.map((it) => `utils/${it}`)
];
possiblePaths$1 = [
	...possiblePaths$1,
	...possiblePaths$1.map((it) => `src/${it}`),
	...possiblePaths$1.map((it) => `app/${it}`)
];
function resolveReferencePath(configDir, refPath) {
	const resolvedPath = path.resolve(configDir, refPath);
	if (refPath.endsWith(".json")) return resolvedPath;
	if (fs.existsSync(resolvedPath)) try {
		if (fs.statSync(resolvedPath).isFile()) return resolvedPath;
	} catch {}
	return path.resolve(configDir, refPath, "tsconfig.json");
}
function getPathAliasesRecursive(tsconfigPath, visited = /* @__PURE__ */ new Set()) {
	if (visited.has(tsconfigPath)) return {};
	visited.add(tsconfigPath);
	if (!fs.existsSync(tsconfigPath)) {
		console.warn(`Referenced tsconfig not found: ${tsconfigPath}`);
		return {};
	}
	try {
		const tsConfig = getTsconfigInfo(void 0, tsconfigPath);
		const { paths = {}, baseUrl = "." } = tsConfig.compilerOptions || {};
		const result = {};
		const configDir = path.dirname(tsconfigPath);
		const obj = Object.entries(paths);
		for (const [alias, aliasPaths] of obj) for (const aliasedPath of aliasPaths) {
			const resolvedBaseUrl = path.resolve(configDir, baseUrl);
			const finalAlias = alias.slice(-1) === "*" ? alias.slice(0, -1) : alias;
			const finalAliasedPath = aliasedPath.slice(-1) === "*" ? aliasedPath.slice(0, -1) : aliasedPath;
			result[finalAlias || ""] = path.join(resolvedBaseUrl, finalAliasedPath);
		}
		if (tsConfig.references) for (const ref of tsConfig.references) {
			const refAliases = getPathAliasesRecursive(resolveReferencePath(configDir, ref.path), visited);
			for (const [alias, aliasPath] of Object.entries(refAliases)) if (!(alias in result)) result[alias] = aliasPath;
		}
		return result;
	} catch (error) {
		console.warn(`Error parsing tsconfig at ${tsconfigPath}: ${error}`);
		return {};
	}
}
function getPathAliases(cwd) {
	let tsConfigPath = path.join(cwd, "tsconfig.json");
	if (!fs.existsSync(tsConfigPath)) tsConfigPath = path.join(cwd, "jsconfig.json");
	if (!fs.existsSync(tsConfigPath)) return null;
	try {
		const result = getPathAliasesRecursive(tsConfigPath);
		addSvelteKitEnvModules(result);
		addCloudflareModules(result);
		return result;
	} catch (error) {
		console.error(error);
		throw new BetterAuthError("Error parsing tsconfig.json");
	}
}
/**
* .tsx files are not supported by Jiti.
*/
const jitiOptions = (cwd) => {
	const alias = getPathAliases(cwd) || {};
	return {
		transformOptions: { babel: { presets: [[babelPresetTypeScript, {
			isTSX: true,
			allExtensions: true
		}], [babelPresetReact, { runtime: "automatic" }]] } },
		extensions: [
			".ts",
			".tsx",
			".js",
			".jsx"
		],
		alias
	};
};
const isDefaultExport = (object) => {
	return typeof object === "object" && object !== null && !Array.isArray(object) && Object.keys(object).length > 0 && "options" in object;
};
async function getConfig({ cwd, configPath, shouldThrowOnError = false }) {
	try {
		let configFile = null;
		if (configPath) {
			let resolvedPath = path.join(cwd, configPath);
			if (existsSync(configPath)) resolvedPath = configPath;
			const { config } = await loadConfig({
				configFile: resolvedPath,
				dotenv: { fileName: [".env", ".env.local"] },
				jitiOptions: jitiOptions(cwd),
				cwd
			});
			if (!("auth" in config) && !isDefaultExport(config)) {
				if (shouldThrowOnError) throw new Error(`Couldn't read your auth config in ${resolvedPath}. Make sure to default export your auth instance or to export as a variable named auth.`);
				console.error(`[#better-auth]: Couldn't read your auth config in ${resolvedPath}. Make sure to default export your auth instance or to export as a variable named auth.`);
				process.exit(1);
			}
			configFile = "auth" in config ? config.auth?.options : config.options;
		}
		if (!configFile) for (const possiblePath of possiblePaths$1) try {
			const { config } = await loadConfig({
				configFile: possiblePath,
				dotenv: { fileName: [".env", ".env.local"] },
				jitiOptions: jitiOptions(cwd),
				cwd
			});
			if (Object.keys(config).length > 0) {
				configFile = config.auth?.options || config.default?.options || null;
				if (!configFile) {
					if (shouldThrowOnError) throw new Error("Couldn't read your auth config. Make sure to default export your auth instance or to export as a variable named auth.");
					console.error("[#better-auth]: Couldn't read your auth config.");
					console.log("");
					console.log("[#better-auth]: Make sure to default export your auth instance or to export as a variable named auth.");
					process.exit(1);
				}
				break;
			}
		} catch (e) {
			if (typeof e === "object" && e && "message" in e && typeof e.message === "string" && e.message.includes("This module cannot be imported from a Client Component module")) {
				if (shouldThrowOnError) throw new Error(`Please remove import 'server-only' from your auth config file temporarily. The CLI cannot resolve the configuration with it included. You can re-add it after running the CLI.`);
				console.error(`Please remove import 'server-only' from your auth config file temporarily. The CLI cannot resolve the configuration with it included. You can re-add it after running the CLI.`);
				process.exit(1);
			}
			if (shouldThrowOnError) throw e;
			console.error("[#better-auth]: Couldn't read your auth config.", e);
			process.exit(1);
		}
		return configFile;
	} catch (e) {
		if (typeof e === "object" && e && "message" in e && typeof e.message === "string" && e.message.includes("This module cannot be imported from a Client Component module")) {
			if (shouldThrowOnError) throw new Error(`Please remove import 'server-only' from your auth config file temporarily. The CLI cannot resolve the configuration with it included. You can re-add it after running the CLI.`);
			console.error(`Please remove import 'server-only' from your auth config file temporarily. The CLI cannot resolve the configuration with it included. You can re-add it after running the CLI.`);
			process.exit(1);
		}
		if (shouldThrowOnError) throw e;
		console.error("Couldn't read your auth config.", e);
		process.exit(1);
	}
}

//#endregion
//#region src/commands/generate.ts
async function generateAction(opts) {
	const options = z$1.object({
		cwd: z$1.string(),
		config: z$1.string().optional(),
		output: z$1.string().optional(),
		y: z$1.boolean().optional(),
		yes: z$1.boolean().optional()
	}).parse(opts);
	const cwd = path.resolve(options.cwd);
	if (!existsSync(cwd)) {
		console.error(`The directory "${cwd}" does not exist.`);
		process.exit(1);
	}
	const config = await getConfig({
		cwd,
		configPath: options.config
	});
	if (!config) {
		console.error("No configuration file found. Add a `auth.ts` file to your project or pass the path to the configuration file using the `--config` flag.");
		return;
	}
	const adapter = await getAdapter(config).catch((e) => {
		console.error(e.message);
		process.exit(1);
	});
	const spinner = yoctoSpinner({ text: "preparing schema..." }).start();
	const schema = await generateSchema({
		adapter,
		file: options.output,
		options: config
	});
	spinner.stop();
	if (!schema.code) {
		console.log("Your schema is already up to date.");
		try {
			await (await createTelemetry(config)).publish({
				type: "cli_generate",
				payload: {
					outcome: "no_changes",
					config: await getTelemetryAuthConfig(config, {
						adapter: adapter.id,
						database: typeof config.database === "function" ? "adapter" : "kysely"
					})
				}
			});
		} catch {}
		process.exit(0);
	}
	if (schema.overwrite) {
		let confirm = options.y || options.yes;
		if (!confirm) confirm = (await prompts({
			type: "confirm",
			name: "confirm",
			message: `The file ${schema.fileName} already exists. Do you want to ${chalk.yellow(`${schema.overwrite ? "overwrite" : "append"}`)} the schema to the file?`
		})).confirm;
		if (confirm) {
			if (!existsSync(path.join(cwd, schema.fileName))) await fs$1.mkdir(path.dirname(path.join(cwd, schema.fileName)), { recursive: true });
			if (schema.overwrite) await fs$1.writeFile(path.join(cwd, schema.fileName), schema.code);
			else await fs$1.appendFile(path.join(cwd, schema.fileName), schema.code);
			console.log(`🚀 Schema was ${schema.overwrite ? "overwritten" : "appended"} successfully!`);
			try {
				await (await createTelemetry(config)).publish({
					type: "cli_generate",
					payload: {
						outcome: schema.overwrite ? "overwritten" : "appended",
						config: await getTelemetryAuthConfig(config)
					}
				});
			} catch {}
			process.exit(0);
		} else {
			console.error("Schema generation aborted.");
			try {
				await (await createTelemetry(config)).publish({
					type: "cli_generate",
					payload: {
						outcome: "aborted",
						config: await getTelemetryAuthConfig(config)
					}
				});
			} catch {}
			process.exit(1);
		}
	}
	if (options.y) {
		console.warn("WARNING: --y is deprecated. Consider -y or --yes");
		options.yes = true;
	}
	let confirm = options.yes;
	if (!confirm) confirm = (await prompts({
		type: "confirm",
		name: "confirm",
		message: `Do you want to generate the schema to ${chalk.yellow(schema.fileName)}?`
	})).confirm;
	if (!confirm) {
		console.error("Schema generation aborted.");
		try {
			await (await createTelemetry(config)).publish({
				type: "cli_generate",
				payload: {
					outcome: "aborted",
					config: await getTelemetryAuthConfig(config)
				}
			});
		} catch {}
		process.exit(1);
	}
	if (!options.output) {
		if (!existsSync(path.dirname(path.join(cwd, schema.fileName)))) await fs$1.mkdir(path.dirname(path.join(cwd, schema.fileName)), { recursive: true });
	}
	await fs$1.writeFile(options.output || path.join(cwd, schema.fileName), schema.code);
	console.log(`🚀 Schema was generated successfully!`);
	try {
		await (await createTelemetry(config)).publish({
			type: "cli_generate",
			payload: {
				outcome: "generated",
				config: await getTelemetryAuthConfig(config)
			}
		});
	} catch {}
	process.exit(0);
}
const generate = new Command("generate").option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).option("--config <config>", "the path to the configuration file. defaults to the first configuration file found.").option("--output <output>", "the file to output to the generated schema").option("-y, --yes", "automatically answer yes to all prompts", false).option("--y", "(deprecated) same as --yes", false).action(generateAction);

//#endregion
//#region src/commands/info.ts
function getSystemInfo() {
	const platform = os.platform();
	const arch = os.arch();
	const version = os.version();
	const release = os.release();
	const cpus = os.cpus();
	const memory = os.totalmem();
	const freeMemory = os.freemem();
	return {
		platform,
		arch,
		version,
		release,
		cpuCount: cpus.length,
		cpuModel: cpus[0]?.model || "Unknown",
		totalMemory: `${(memory / 1024 / 1024 / 1024).toFixed(2)} GB`,
		freeMemory: `${(freeMemory / 1024 / 1024 / 1024).toFixed(2)} GB`
	};
}
function getNodeInfo() {
	return {
		version: process.version,
		env: process.env.NODE_ENV || "development"
	};
}
function getPackageManager() {
	const userAgent = process.env.npm_config_user_agent || "";
	if (userAgent.includes("yarn")) return {
		name: "yarn",
		version: getVersion$1("yarn")
	};
	if (userAgent.includes("pnpm")) return {
		name: "pnpm",
		version: getVersion$1("pnpm")
	};
	if (userAgent.includes("bun")) return {
		name: "bun",
		version: getVersion$1("bun")
	};
	return {
		name: "npm",
		version: getVersion$1("npm")
	};
}
function getVersion$1(command) {
	try {
		return execSync(`${command} --version`, { encoding: "utf8" }).trim();
	} catch {
		return "Not installed";
	}
}
function getFrameworkInfo(projectRoot) {
	const packageJsonPath = path.join(projectRoot, "package.json");
	if (!existsSync(packageJsonPath)) return null;
	try {
		const packageJson = JSON.parse(readFileSync(packageJsonPath, "utf8"));
		const deps = {
			...packageJson.dependencies,
			...packageJson.devDependencies
		};
		const frameworks = {
			next: deps["next"],
			react: deps["react"],
			vue: deps["vue"],
			nuxt: deps["nuxt"],
			svelte: deps["svelte"],
			"@sveltejs/kit": deps["@sveltejs/kit"],
			express: deps["express"],
			fastify: deps["fastify"],
			hono: deps["hono"],
			"react-router": deps["react-router"],
			astro: deps["astro"],
			solid: deps["solid-js"],
			qwik: deps["@builder.io/qwik"]
		};
		const installedFrameworks = Object.entries(frameworks).filter(([_, version]) => version).map(([name, version]) => ({
			name,
			version
		}));
		return installedFrameworks.length > 0 ? installedFrameworks : null;
	} catch {
		return null;
	}
}
function getDatabaseInfo(projectRoot) {
	const packageJsonPath = path.join(projectRoot, "package.json");
	if (!existsSync(packageJsonPath)) return null;
	try {
		const packageJson = JSON.parse(readFileSync(packageJsonPath, "utf8"));
		const deps = {
			...packageJson.dependencies,
			...packageJson.devDependencies
		};
		const databases = {
			"better-sqlite3": deps["better-sqlite3"],
			"@libsql/client": deps["@libsql/client"],
			"@libsql/kysely-libsql": deps["@libsql/kysely-libsql"],
			mysql2: deps["mysql2"],
			pg: deps["pg"],
			postgres: deps["postgres"],
			"@prisma/client": deps["@prisma/client"],
			drizzle: deps["drizzle-orm"],
			kysely: deps["kysely"],
			mongodb: deps["mongodb"],
			"@neondatabase/serverless": deps["@neondatabase/serverless"],
			"@vercel/postgres": deps["@vercel/postgres"],
			"@planetscale/database": deps["@planetscale/database"]
		};
		const installedDatabases = Object.entries(databases).filter(([_, version]) => version).map(([name, version]) => ({
			name,
			version
		}));
		return installedDatabases.length > 0 ? installedDatabases : null;
	} catch {
		return null;
	}
}
function sanitizeBetterAuthConfig(config) {
	if (!config) return null;
	const sanitized = JSON.parse(JSON.stringify(config));
	const sensitiveKeys = [
		"secret",
		"clientSecret",
		"clientId",
		"authToken",
		"apiKey",
		"apiSecret",
		"privateKey",
		"publicKey",
		"password",
		"token",
		"webhook",
		"connectionString",
		"databaseUrl",
		"databaseURL",
		"TURSO_AUTH_TOKEN",
		"TURSO_DATABASE_URL",
		"MYSQL_DATABASE_URL",
		"DATABASE_URL",
		"POSTGRES_URL",
		"MONGODB_URI",
		"stripeKey",
		"stripeWebhookSecret"
	];
	const allowedKeys = [
		"baseURL",
		"callbackURL",
		"redirectURL",
		"trustedOrigins",
		"appName"
	];
	function redactSensitive(obj, parentKey) {
		if (typeof obj !== "object" || obj === null) {
			if (parentKey && typeof obj === "string" && obj.length > 0) {
				if (allowedKeys.some((allowed) => parentKey.toLowerCase() === allowed.toLowerCase())) return obj;
				const lowerKey = parentKey.toLowerCase();
				if (sensitiveKeys.some((key) => {
					const lowerSensitiveKey = key.toLowerCase();
					return lowerKey === lowerSensitiveKey || lowerKey.endsWith(lowerSensitiveKey);
				})) return "[REDACTED]";
			}
			return obj;
		}
		if (Array.isArray(obj)) return obj.map((item) => redactSensitive(item, parentKey));
		const result = {};
		for (const [key, value] of Object.entries(obj)) {
			if (allowedKeys.some((allowed) => key.toLowerCase() === allowed.toLowerCase())) {
				result[key] = value;
				continue;
			}
			const lowerKey = key.toLowerCase();
			if (sensitiveKeys.some((sensitiveKey) => {
				const lowerSensitiveKey = sensitiveKey.toLowerCase();
				return lowerKey === lowerSensitiveKey || lowerKey.endsWith(lowerSensitiveKey);
			})) if (typeof value === "string" && value.length > 0) result[key] = "[REDACTED]";
			else if (typeof value === "object" && value !== null) result[key] = redactSensitive(value, key);
			else result[key] = value;
			else result[key] = redactSensitive(value, key);
		}
		return result;
	}
	if (sanitized.database) {
		if (typeof sanitized.database === "string") sanitized.database = "[REDACTED]";
		else if (sanitized.database.url) sanitized.database.url = "[REDACTED]";
		if (sanitized.database.authToken) sanitized.database.authToken = "[REDACTED]";
	}
	if (sanitized.socialProviders) {
		for (const provider in sanitized.socialProviders) if (sanitized.socialProviders[provider]) sanitized.socialProviders[provider] = redactSensitive(sanitized.socialProviders[provider], provider);
	}
	if (sanitized.emailAndPassword?.sendResetPassword) sanitized.emailAndPassword.sendResetPassword = "[Function]";
	if (sanitized.emailVerification?.sendVerificationEmail) sanitized.emailVerification.sendVerificationEmail = "[Function]";
	if (sanitized.plugins && Array.isArray(sanitized.plugins)) sanitized.plugins = sanitized.plugins.map((plugin) => {
		if (typeof plugin === "function") return "[Plugin Function]";
		if (plugin && typeof plugin === "object") return {
			name: plugin.id || plugin.name || "unknown",
			config: redactSensitive(plugin.config || plugin)
		};
		return plugin;
	});
	return redactSensitive(sanitized);
}
async function getBetterAuthInfo(projectRoot, configPath, suppressLogs = false) {
	try {
		const originalLog = console.log;
		const originalWarn = console.warn;
		const originalError = console.error;
		if (suppressLogs) {
			console.log = () => {};
			console.warn = () => {};
			console.error = () => {};
		}
		try {
			const config = await getConfig({
				cwd: projectRoot,
				configPath,
				shouldThrowOnError: true
			});
			const packageInfo = await getPackageInfo();
			return {
				version: packageInfo.dependencies?.["better-auth"] || packageInfo.devDependencies?.["better-auth"] || packageInfo.peerDependencies?.["better-auth"] || packageInfo.optionalDependencies?.["better-auth"] || "Unknown",
				config: sanitizeBetterAuthConfig(config)
			};
		} finally {
			if (suppressLogs) {
				console.log = originalLog;
				console.warn = originalWarn;
				console.error = originalError;
			}
		}
	} catch (error) {
		return {
			version: "Unknown",
			config: null,
			error: error instanceof Error ? error.message : "Failed to load Better Auth config"
		};
	}
}
function formatOutput(data, indent = 0) {
	const spaces = " ".repeat(indent);
	if (data === null || data === void 0) return `${spaces}${chalk.gray("N/A")}`;
	if (typeof data === "string" || typeof data === "number" || typeof data === "boolean") return `${spaces}${data}`;
	if (Array.isArray(data)) {
		if (data.length === 0) return `${spaces}${chalk.gray("[]")}`;
		return data.map((item) => formatOutput(item, indent)).join("\n");
	}
	if (typeof data === "object") {
		const entries = Object.entries(data);
		if (entries.length === 0) return `${spaces}${chalk.gray("{}")}`;
		return entries.map(([key, value]) => {
			if (typeof value === "object" && value !== null && !Array.isArray(value)) return `${spaces}${chalk.cyan(key)}:\n${formatOutput(value, indent + 2)}`;
			return `${spaces}${chalk.cyan(key)}: ${formatOutput(value, 0)}`;
		}).join("\n");
	}
	return `${spaces}${JSON.stringify(data)}`;
}
const info = new Command("info").description("Display system and Better Auth configuration information").option("--cwd <cwd>", "The working directory", process.cwd()).option("--config <config>", "Path to the Better Auth configuration file").option("-j, --json", "Output as JSON").option("-c, --copy", "Copy output to clipboard (requires pbcopy/xclip)").action(async (options) => {
	const projectRoot = path.resolve(options.cwd || process.cwd());
	const systemInfo = getSystemInfo();
	const nodeInfo = getNodeInfo();
	const packageManager = getPackageManager();
	const frameworks = getFrameworkInfo(projectRoot);
	const databases = getDatabaseInfo(projectRoot);
	const betterAuthInfo = await getBetterAuthInfo(projectRoot, options.config, options.json);
	const fullInfo = {
		system: systemInfo,
		node: nodeInfo,
		packageManager,
		frameworks,
		databases,
		betterAuth: betterAuthInfo
	};
	if (options.json) {
		const jsonOutput = JSON.stringify(fullInfo, null, 2);
		console.log(jsonOutput);
		if (options.copy) try {
			const platform = os.platform();
			if (platform === "darwin") {
				execSync("pbcopy", { input: jsonOutput });
				console.log(chalk.green("\n✓ Copied to clipboard"));
			} else if (platform === "linux") {
				execSync("xclip -selection clipboard", { input: jsonOutput });
				console.log(chalk.green("\n✓ Copied to clipboard"));
			} else if (platform === "win32") {
				execSync("clip", { input: jsonOutput });
				console.log(chalk.green("\n✓ Copied to clipboard"));
			}
		} catch {
			console.log(chalk.yellow("\n⚠ Could not copy to clipboard"));
		}
		return;
	}
	console.log(chalk.bold("\n📊 Better Auth System Information\n"));
	console.log(chalk.gray("=".repeat(50)));
	console.log(chalk.bold.white("\n🖥️  System Information:"));
	console.log(formatOutput(systemInfo, 2));
	console.log(chalk.bold.white("\n📦 Node.js:"));
	console.log(formatOutput(nodeInfo, 2));
	console.log(chalk.bold.white("\n📦 Package Manager:"));
	console.log(formatOutput(packageManager, 2));
	if (frameworks) {
		console.log(chalk.bold.white("\n🚀 Frameworks:"));
		console.log(formatOutput(frameworks, 2));
	}
	if (databases) {
		console.log(chalk.bold.white("\n💾 Database Clients:"));
		console.log(formatOutput(databases, 2));
	}
	console.log(chalk.bold.white("\n🔐 Better Auth:"));
	if (betterAuthInfo.error) console.log(`  ${chalk.red("Error:")} ${betterAuthInfo.error}`);
	else {
		console.log(`  ${chalk.cyan("Version")}: ${betterAuthInfo.version}`);
		if (betterAuthInfo.config) {
			console.log(`  ${chalk.cyan("Configuration")}:`);
			console.log(formatOutput(betterAuthInfo.config, 4));
		}
	}
	console.log(chalk.gray("\n" + "=".repeat(50)));
	console.log(chalk.gray("\n💡 Tip: Use --json flag for JSON output"));
	console.log(chalk.gray("💡 Use --copy flag to copy output to clipboard"));
	console.log(chalk.gray("💡 When reporting issues, include this information\n"));
	if (options.copy) {
		const textOutput = `
Better Auth System Information
==============================

System Information:
${JSON.stringify(systemInfo, null, 2)}

Node.js:
${JSON.stringify(nodeInfo, null, 2)}

Package Manager:
${JSON.stringify(packageManager, null, 2)}

Frameworks:
${JSON.stringify(frameworks, null, 2)}

Database Clients:
${JSON.stringify(databases, null, 2)}

Better Auth:
${JSON.stringify(betterAuthInfo, null, 2)}
`;
		try {
			const platform = os.platform();
			if (platform === "darwin") {
				execSync("pbcopy", { input: textOutput });
				console.log(chalk.green("✓ Copied to clipboard"));
			} else if (platform === "linux") {
				execSync("xclip -selection clipboard", { input: textOutput });
				console.log(chalk.green("✓ Copied to clipboard"));
			} else if (platform === "win32") {
				execSync("clip", { input: textOutput });
				console.log(chalk.green("✓ Copied to clipboard"));
			}
		} catch {
			console.log(chalk.yellow("⚠ Could not copy to clipboard"));
		}
	}
});

//#endregion
//#region src/utils/check-package-managers.ts
async function checkPackageManagers() {
	return {
		hasPnpm: await getVersion("pnpm"),
		hasBun: await getVersion("bun"),
		hasYarn: await getVersion("yarn")
	};
}
const PACKAGE_MANAGER = [
	"npm",
	"yarn",
	"pnpm",
	"bun"
];
async function detectPackageManager(cwd, packageJson) {
	const monorepoRoot = await findMonorepoRoot(cwd);
	for (const strategy of [
		envStrategy,
		packageJsonStrategy$1,
		lockFileStrategy,
		configStrategy,
		cliStrategy
	]) {
		const result = await strategy({
			cwd: monorepoRoot ?? cwd,
			packageJson
		});
		if (result !== null && PACKAGE_MANAGER.includes(result.packageManager.toLowerCase())) return result;
	}
	return { packageManager: "npm" };
}
const envStrategy = () => {
	const userAgent = env.npm_config_user_agent;
	if (!userAgent) return null;
	const pmSpec = userAgent.split(" ")[0];
	const separatorPos = pmSpec.lastIndexOf("/");
	return {
		packageManager: pmSpec.substring(0, separatorPos),
		version: pmSpec.substring(separatorPos + 1)
	};
};
const lockFileStrategy = ({ cwd }) => {
	if (existsSync(join(cwd, "package-lock.json"))) return { packageManager: "npm" };
	if (existsSync(join(cwd, "yarn.lock"))) return { packageManager: "yarn" };
	if (existsSync(join(cwd, "pnpm-lock.yaml"))) return { packageManager: "pnpm" };
	if (existsSync(join(cwd, "bun.lock")) || existsSync(join(cwd, "bun.lockb"))) return { packageManager: "bun" };
	return null;
};
const packageJsonStrategy$1 = ({ packageJson }) => {
	const [packageManager, version] = packageJson.packageManager?.split("@", 2) ?? [];
	if (packageManager && PACKAGE_MANAGER.includes(packageManager.toLowerCase())) return {
		packageManager,
		version
	};
	return null;
};
const configStrategy = ({ cwd, packageJson }) => {
	if (typeof packageJson.workspaces === "object") {
		if ("nohoist" in packageJson.workspaces) return { packageManager: "yarn" };
		if ("catalog" in packageJson.workspaces) return { packageManager: "bun" };
	}
	if (typeof packageJson.pnpm !== "undefined" || existsSync(join(cwd, "pnpm-workspace.yaml"))) return { packageManager: "pnpm" };
	if (existsSync(join(cwd, ".yarnrc.yml")) || existsSync(join(cwd, ".yarnrc"))) return { packageManager: "yarn" };
	if (existsSync(join(cwd, "bunfig.toml"))) return { packageManager: "bun" };
	return null;
};
const cliStrategy = async ({ cwd }) => {
	const { hasBun, hasPnpm, hasYarn } = await checkPackageManagers();
	if (hasBun) return { packageManager: "bun" };
	if (hasPnpm) return { packageManager: "pnpm" };
	if (hasYarn) return { packageManager: "yarn" };
	return null;
};
function getPkgManagerStr({ packageManager, version }) {
	if (!version) return packageManager;
	return `${packageManager}@${version}`;
}
async function getVersion(pkgManager) {
	return await new Promise((resolve) => {
		exec(`${pkgManager} -v`, (err, stdout) => {
			if (err) {
				resolve(null);
				return;
			}
			resolve(stdout.trim());
		});
	});
}

//#endregion
//#region src/utils/config-paths.ts
let possiblePaths = [
	"auth.ts",
	"auth.tsx",
	"auth.js",
	"auth.jsx",
	"auth.server.js",
	"auth.server.ts",
	"auth/index.ts",
	"auth/index.tsx",
	"auth/index.js",
	"auth/index.jsx",
	"auth/index.server.js",
	"auth/index.server.ts"
];
possiblePaths = [
	...possiblePaths,
	...possiblePaths.map((it) => `lib/server/${it}`),
	...possiblePaths.map((it) => `server/auth/${it}`),
	...possiblePaths.map((it) => `server/${it}`),
	...possiblePaths.map((it) => `auth/${it}`),
	...possiblePaths.map((it) => `lib/${it}`),
	...possiblePaths.map((it) => `utils/${it}`)
];
possiblePaths = [
	...possiblePaths,
	...possiblePaths.map((it) => `src/${it}`),
	...possiblePaths.map((it) => `app/${it}`)
];
const possibleAuthConfigPaths = possiblePaths;
let _possibleClientConfigPaths = [
	"auth-client.ts",
	"auth-client.tsx",
	"auth-client.js",
	"auth-client.jsx",
	"auth-client.server.js",
	"auth-client.server.ts",
	"auth-client/index.ts",
	"auth-client/index.tsx",
	"auth-client/index.js",
	"auth-client/index.jsx",
	"auth-client/index.server.js",
	"auth-client/index.server.ts"
];
_possibleClientConfigPaths = [
	..._possibleClientConfigPaths,
	..._possibleClientConfigPaths.map((it) => `lib/server/${it}`),
	..._possibleClientConfigPaths.map((it) => `server/auth/${it}`),
	..._possibleClientConfigPaths.map((it) => `server/${it}`),
	..._possibleClientConfigPaths.map((it) => `auth/${it}`),
	..._possibleClientConfigPaths.map((it) => `lib/${it}`),
	..._possibleClientConfigPaths.map((it) => `utils/${it}`)
];
_possibleClientConfigPaths = [
	..._possibleClientConfigPaths,
	..._possibleClientConfigPaths.map((it) => `src/${it}`),
	..._possibleClientConfigPaths.map((it) => `app/${it}`)
];
const possibleClientConfigPaths = _possibleClientConfigPaths;

//#endregion
//#region src/utils/install-dependencies.ts
const flagsMap = {
	npm: {
		dev: "--save-dev",
		optional: "--save-optional"
	},
	pnpm: {
		dev: "--save-dev",
		peer: "--save-peer",
		optional: "--save-optional",
		catalog: (name) => {
			if (name) return `--save-catalog-name ${name}`;
			return "--save-catalog";
		}
	},
	bun: {
		dev: "--dev",
		peer: "--peer",
		optional: "--optional"
	},
	yarn: {
		dev: "--dev",
		peer: "--peer",
		optional: "--optional"
	}
};
function installDependencies({ dependencies, packageManager, cwd, type = "prod", catalogName }) {
	let installCommand;
	const flags = [];
	switch (packageManager) {
		case "npm":
			installCommand = "npm install";
			flags.push("--force");
			break;
		case "pnpm":
			installCommand = "pnpm add";
			break;
		case "bun":
			installCommand = "bun install";
			break;
		case "yarn":
			installCommand = "yarn install";
			break;
		default: throw new Error("Invalid package manager");
	}
	const flagMap = flagsMap[packageManager];
	if (type === "catalog") if ("catalog" in flagMap) {
		const catalogFlag = flagMap["catalog"];
		flags.push(catalogFlag(catalogName));
	} else throw new Error(`Catalog flag is not supported by "${packageManager}"`);
	else {
		const flag = flagMap?.[type];
		if (flag) flags.push(flag);
	}
	const command = `${installCommand}${flags.length > 0 ? ` ${flags.join(" ")}` : ""} ${Array.isArray(dependencies) ? dependencies.join(" ") : dependencies}`;
	return new Promise((resolve, reject) => {
		exec(command, { cwd }, (error, stdout, stderr) => {
			if (error) {
				reject(new Error(stderr));
				return;
			}
			resolve(true);
		});
	});
}

//#endregion
//#region src/commands/init/configs/frameworks.config.ts
const FRAMEWORKS = [
	{
		name: "Astro",
		id: "astro",
		dependency: "astro",
		authClient: { importPath: "better-auth/react" },
		routeHandler: {
			path: "pages/api/auth/[...all].ts",
			code: `import { auth } from "~/auth";
import type { APIRoute } from "astro";

export const ALL: APIRoute = async (ctx) => {
	// If you want to use rate limiting, make sure to set the 'x-forwarded-for' header to the request headers from the context
	// ctx.request.headers.set("x-forwarded-for", ctx.clientAddress);
	return auth.handler(ctx.request);
};`
		},
		configPaths: [
			"astro.config.mjs",
			"astro.config.ts",
			"astro.config.js",
			"astro.config.cjs"
		]
	},
	{
		name: "Remix",
		id: "remix",
		dependency: "@remix-run/server-runtime",
		authClient: { importPath: "better-auth/react" },
		routeHandler: {
			path: "app/lib/auth.server.ts",
			code: `import { betterAuth } from "better-auth"

export const auth = betterAuth({
    database: {
        provider: "postgres", //change this to your database provider
        url: process.env.DATABASE_URL, // path to your database or connection string
    }
})`
		},
		configPaths: ["remix.config.js"]
	},
	{
		name: "React Router v7",
		id: "react-router-v7",
		dependency: "react-router",
		authClient: { importPath: "better-auth/react" },
		routeHandler: {
			path: "app/lib/auth.server.ts",
			code: `import { betterAuth } from "better-auth"

export const auth = betterAuth({
    database: {
        provider: "postgres", //change this to your database provider
        url: process.env.DATABASE_URL, // path to your database or connection string
    }
})`
		},
		configPaths: ["react-router.config.ts"]
	},
	{
		name: "Next.js",
		id: "next",
		dependency: "next",
		authClient: { importPath: "better-auth/react" },
		routeHandler: {
			path: "api/auth/[...all]/route.ts",
			code: `import { auth } from "@/lib/auth";
import { toNextJsHandler } from "better-auth/next-js";
export const { GET, POST } = toNextJsHandler(auth.handler);`
		},
		configPaths: [
			"next.config.js",
			"next.config.ts",
			"next.config.mjs",
			".next/server/next.config.js",
			".next/server/next.config.ts",
			".next/server/next.config.mjs"
		]
	},
	{
		name: "Nuxt",
		id: "nuxt",
		dependency: "nuxt",
		authClient: { importPath: "better-auth/vue" },
		routeHandler: {
			path: "server/api/auth/[...all].ts",
			code: `import { auth } from "~/lib/auth"; // import your auth config

export default defineEventHandler((event) => {
	return auth.handler(toWebRequest(event));
});`
		},
		configPaths: [
			"nuxt.config.js",
			"nuxt.config.ts",
			"nuxt.config.mjs",
			"nuxt.config.cjs"
		]
	},
	{
		name: "SvelteKit",
		id: "sveltekit",
		dependency: "@sveltejs/kit",
		authClient: { importPath: "better-auth/svelte" },
		routeHandler: {
			path: `hooks.server.ts`,
			code: `import { auth } from "$lib/auth";
import { svelteKitHandler } from "better-auth/svelte-kit";
import { building } from "$app/environment";

export async function handle({ event, resolve }) {
  return svelteKitHandler({ event, resolve, auth, building });
}`
		},
		configPaths: [
			"svelte.config.js",
			"svelte.config.ts",
			"svelte.config.mjs",
			"svelte.config.cjs"
		]
	},
	{
		name: "Solid Start",
		id: "solid-start",
		dependency: "solid-start",
		authClient: { importPath: "better-auth/solid" },
		routeHandler: {
			path: `routes/api/auth/*auth.ts`,
			code: `import { auth } from "~/lib/auth";
import { toSolidStartHandler } from "better-auth/solid-start";

export const { GET, POST } = toSolidStartHandler(auth);`
		},
		configPaths: ["app.config.ts"]
	},
	{
		name: "Tanstack Start",
		id: "tanstack-start",
		dependency: "tanstack-start",
		authClient: { importPath: "better-auth/react" },
		routeHandler: {
			path: `src/routes/api/auth/$.ts`,
			code: `import { auth } from '@/lib/auth'
import { createFileRoute } from '@tanstack/react-router'

export const Route = createFileRoute('/api/auth/$')({
  server: {
    handlers: {
      GET: ({ request }) => {
        return auth.handler(request)
      },
      POST: ({ request }) => {
        return auth.handler(request)
      },
    },
  },
})`
		},
		configPaths: null
	},
	{
		name: "Hono",
		id: "hono",
		dependency: "hono",
		authClient: null,
		routeHandler: null,
		configPaths: null
	},
	{
		name: "Fastify",
		id: "fastify",
		dependency: "fastify",
		authClient: null,
		routeHandler: null,
		configPaths: null
	},
	{
		name: "Express",
		id: "express",
		dependency: "express",
		authClient: null,
		routeHandler: null,
		configPaths: null
	},
	{
		name: "Elysia",
		id: "elysia",
		dependency: "elysia",
		authClient: null,
		routeHandler: null,
		configPaths: null
	},
	{
		name: "Nitro",
		id: "nitro",
		dependency: "nitro",
		authClient: null,
		routeHandler: null,
		configPaths: ["nitro.config.ts"]
	}
];

//#endregion
//#region src/commands/init/configs/social-providers.config.ts
const SOCIAL_PROVIDERS = [
	"apple",
	"atlassian",
	"cognito",
	"discord",
	"dropbox",
	"facebook",
	"figma",
	"github",
	"gitlab",
	"google",
	"huggingface",
	"kakao",
	"kick",
	"line",
	"linear",
	"linkedin",
	"microsoft",
	"naver",
	"notion",
	"paybin",
	"paypal",
	"polar",
	"reddit",
	"roblox",
	"salesforce",
	"slack",
	"spotify",
	"tiktok",
	"twitch",
	"twitter",
	"vercel",
	"vk",
	"zoom"
];
/**
* Configuration for each social provider specifying what options are required
*/
const SOCIAL_PROVIDER_CONFIGS = {
	apple: { options: [{
		name: "clientId",
		envVar: "APPLE_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "APPLE_CLIENT_SECRET"
	}] },
	atlassian: { options: [{
		name: "clientId",
		envVar: "ATLASSIAN_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "ATLASSIAN_CLIENT_SECRET"
	}] },
	cognito: { options: [
		{
			name: "clientId",
			envVar: "COGNITO_CLIENT_ID"
		},
		{
			name: "domain",
			envVar: "COGNITO_DOMAIN"
		},
		{
			name: "region",
			envVar: "COGNITO_REGION"
		},
		{
			name: "userPoolId",
			envVar: "COGNITO_USERPOOL_ID"
		}
	] },
	discord: { options: [{
		name: "clientId",
		envVar: "DISCORD_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "DISCORD_CLIENT_SECRET"
	}] },
	dropbox: { options: [{
		name: "clientId",
		envVar: "DROPBOX_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "DROPBOX_CLIENT_SECRET"
	}] },
	facebook: { options: [{
		name: "clientId",
		envVar: "FACEBOOK_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "FACEBOOK_CLIENT_SECRET"
	}] },
	figma: { options: [{
		name: "clientId",
		envVar: "FIGMA_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "FIGMA_CLIENT_SECRET"
	}] },
	github: { options: [{
		name: "clientId",
		envVar: "GITHUB_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "GITHUB_CLIENT_SECRET"
	}] },
	gitlab: { options: [{
		name: "clientId",
		envVar: "GITLAB_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "GITLAB_CLIENT_SECRET"
	}] },
	google: { options: [{
		name: "clientId",
		envVar: "GOOGLE_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "GOOGLE_CLIENT_SECRET"
	}] },
	huggingface: { options: [{
		name: "clientId",
		envVar: "HUGGINGFACE_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "HUGGINGFACE_CLIENT_SECRET"
	}] },
	kakao: { options: [{
		name: "clientId",
		envVar: "KAKAO_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "KAKAO_CLIENT_SECRET"
	}] },
	kick: { options: [{
		name: "clientId",
		envVar: "KICK_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "KICK_CLIENT_SECRET"
	}] },
	line: { options: [{
		name: "clientId",
		envVar: "LINE_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "LINE_CLIENT_SECRET"
	}] },
	linear: { options: [{
		name: "clientId",
		envVar: "LINEAR_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "LINEAR_CLIENT_SECRET"
	}] },
	linkedin: { options: [{
		name: "clientId",
		envVar: "LINKEDIN_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "LINKEDIN_CLIENT_SECRET"
	}] },
	microsoft: { options: [{
		name: "clientId",
		envVar: "MICROSOFT_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "MICROSOFT_CLIENT_SECRET"
	}] },
	naver: { options: [{
		name: "clientId",
		envVar: "NAVER_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "NAVER_CLIENT_SECRET"
	}] },
	notion: { options: [{
		name: "clientId",
		envVar: "NOTION_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "NOTION_CLIENT_SECRET"
	}] },
	paybin: { options: [{
		name: "clientId",
		envVar: "PAYBIN_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "PAYBIN_CLIENT_SECRET"
	}] },
	paypal: { options: [{
		name: "clientId",
		envVar: "PAYPAL_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "PAYPAL_CLIENT_SECRET"
	}] },
	polar: { options: [{
		name: "clientId",
		envVar: "POLAR_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "POLAR_CLIENT_SECRET"
	}] },
	reddit: { options: [{
		name: "clientId",
		envVar: "REDDIT_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "REDDIT_CLIENT_SECRET"
	}] },
	roblox: { options: [{
		name: "clientId",
		envVar: "ROBLOX_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "ROBLOX_CLIENT_SECRET"
	}] },
	salesforce: { options: [{
		name: "clientId",
		envVar: "SALESFORCE_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "SALESFORCE_CLIENT_SECRET"
	}] },
	slack: { options: [{
		name: "clientId",
		envVar: "SLACK_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "SLACK_CLIENT_SECRET"
	}] },
	spotify: { options: [{
		name: "clientId",
		envVar: "SPOTIFY_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "SPOTIFY_CLIENT_SECRET"
	}] },
	tiktok: { options: [{
		name: "clientKey",
		envVar: "TIKTOK_CLIENT_KEY"
	}, {
		name: "clientSecret",
		envVar: "TIKTOK_CLIENT_SECRET"
	}] },
	twitch: { options: [{
		name: "clientId",
		envVar: "TWITCH_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "TWITCH_CLIENT_SECRET"
	}] },
	twitter: { options: [{
		name: "clientId",
		envVar: "TWITTER_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "TWITTER_CLIENT_SECRET"
	}] },
	vercel: { options: [{
		name: "clientId",
		envVar: "VERCEL_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "VERCEL_CLIENT_SECRET"
	}] },
	vk: { options: [{
		name: "clientId",
		envVar: "VK_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "VK_CLIENT_SECRET"
	}] },
	zoom: { options: [{
		name: "clientId",
		envVar: "ZOOM_CLIENT_ID"
	}, {
		name: "clientSecret",
		envVar: "ZOOM_CLIENT_SECRET"
	}] }
};

//#endregion
//#region src/commands/init/utility/format.ts
const formatCode = async (code) => {
	return await format(code, { parser: "typescript" });
};

//#endregion
//#region src/commands/init/utility/imports.ts
/**
* Helper function to create an import object.
*/
const createImport = ({ name, alias, asType }) => {
	return {
		name,
		alias: alias ?? null,
		asType: asType ?? false
	};
};
/**
* Converts an import object to a string. This is specifically for the variables in the import.
* For the full import statement, use the `getImportString` function.
*/
const getImportVariableString = (import_) => {
	const alias = import_.alias ? ` as ${import_.alias}` : "";
	return `${import_.asType ? "type " : ""}${import_.name}${alias}`.trim();
};
/**
* Takes a collection of imports and returns a string of import statements.
*/
const getImportString = async (imports) => {
	const groupedImports = groupImports(imports);
	let importString = "";
	for (const { imports, path, isNamedImport } of groupedImports) {
		const vars = isNamedImport ? getImportVariableString(imports) : `{ ${imports.map(getImportVariableString).join(", ")} }`;
		importString += `import ${vars} from "${path}";\n`;
	}
	return (await formatCode(importString)).trim();
};
/**
* Takes a collection of imports and groups them by path.
*/
const groupImports = (imports) => {
	const result = [];
	for (const import_ of imports) {
		if (import_.isNamedImport) {
			result.push(import_);
			continue;
		}
		const existingIndex = result.findIndex((x) => x.path === import_.path && !x.isNamedImport);
		if (existingIndex !== -1) {
			result[existingIndex].imports.push(...import_.imports);
			continue;
		}
		result.push(import_);
	}
	return result.sort((a, b) => {
		if (a.isNamedImport && !b.isNamedImport) return 1;
		if (!a.isNamedImport && b.isNamedImport) return -1;
		return a.path.localeCompare(b.path);
	});
};

//#endregion
//#region src/commands/init/configs/temp-plugins.config.ts
const tempPluginsConfig = {
	twoFactor: {
		displayName: "Two Factor",
		auth: {
			function: "twoFactor",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "twoFactor" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "two-factor-issuer",
					question: "What is the issuer for the two factor authentication?",
					description: "The issuer for the two factor authentication.",
					defaultValue: "My App",
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "issuer",
						schema: z$1.coerce.string().optional()
					}
				},
				{
					flag: "skip-verification-on-enable",
					question: "Skip verification on enable two factor authentication?",
					description: "Skip verification on enable two factor authentication.",
					defaultValue: false,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "skipVerificationOnEnable",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "totp",
					description: "The number of digits for the TOTP code.",
					defaultValue: "My App",
					skip: "prompt",
					isNestedObject: [{
						flag: "totp-otp-digits",
						question: "What is the number of digits for the TOTP code?",
						description: "The number of digits for the TOTP code.",
						defaultValue: 6,
						skip: "prompt",
						argument: {
							index: 0,
							isProperty: "digits",
							schema: z$1.coerce.number().positive().optional()
						}
					}, {
						flag: "totp-otp-period",
						question: "What is the period for the TOTP code?",
						description: "The period for the TOTP code.",
						defaultValue: 30,
						skip: "prompt",
						argument: {
							index: 0,
							isProperty: "period",
							schema: z$1.coerce.number().positive().optional()
						}
					}],
					argument: {
						index: 0,
						isProperty: "totp"
					}
				},
				{
					flag: "otp",
					description: "The options for the OTP code.",
					defaultValue: 3,
					skip: "prompt",
					isNestedObject: [{
						flag: "otp-period",
						question: "What is the period for the OTP code?",
						description: "The period for the OTP code.",
						defaultValue: 3,
						skip: "prompt",
						argument: {
							index: 0,
							isProperty: "period",
							schema: z$1.coerce.number().positive().optional()
						}
					}, {
						flag: "otp-store-otp",
						question: "How do you want to store the OTP code?",
						description: "The function to store the OTP code.",
						defaultValue: "storeOTP",
						skip: "prompt",
						isSelectOptions: [
							{
								value: "plain",
								label: "Plain text"
							},
							{
								value: "encrypted",
								label: "Encrypted"
							},
							{
								value: "hashed",
								label: "Hashed"
							}
						],
						argument: {
							index: 0,
							isProperty: "storeOTP",
							schema: z$1.enum([
								"plain",
								"encrypted",
								"hashed"
							]).optional()
						}
					}],
					argument: {
						index: 0,
						isProperty: "otp"
					}
				},
				{
					flag: "backup-code",
					description: "The options for the backup code.",
					skip: "prompt",
					isNestedObject: [{
						flag: "backup-code-amount",
						question: "What is the amount of backup codes to generate?",
						description: "The amount of backup codes to generate.",
						defaultValue: 10,
						skip: "prompt",
						argument: {
							index: 0,
							isProperty: "amount",
							schema: z$1.coerce.number().positive().optional()
						}
					}, {
						flag: "backup-code-length",
						question: "What is the length of the backup codes?",
						description: "The length of the backup codes.",
						defaultValue: 10,
						skip: "prompt",
						argument: {
							index: 0,
							isProperty: "length",
							schema: z$1.coerce.number().positive().optional()
						}
					}],
					argument: {
						index: 0,
						isProperty: "backupCodeOptions"
					}
				},
				{
					flag: "two-factor-schema",
					argument: {
						index: 0,
						isProperty: "schema"
					},
					description: "The schema for the two factor plugin.",
					skip: "prompt",
					isNestedObject: [{
						flag: "two-factor-table",
						question: "What is the name of the two factor table?",
						description: "The name of the two factor table.",
						defaultValue: "twoFactor",
						skip: "prompt",
						argument: {
							index: 0,
							isProperty: "twoFactorTable",
							schema: z$1.coerce.string().optional()
						}
					}]
				}
			]
		},
		authClient: {
			function: "twoFactorClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "twoFactorClient" })],
				isNamedImport: false
			}]
		}
	},
	username: {
		displayName: "Username",
		auth: {
			function: "username",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "username" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "username-max-username-length",
					question: "What is the maximum length of the username?",
					description: "The maximum length of the username.",
					isNumber: true,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "maxUsernameLength",
						schema: z$1.coerce.number().min(0).positive().optional()
					}
				},
				{
					flag: "username-min-username-length",
					question: "What is the minimum length of the username?",
					description: "The minimum length of the username.",
					isNumber: true,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "minUsernameLength",
						schema: z$1.coerce.number().min(0).positive().optional()
					}
				},
				{
					flag: "username-validation-order",
					description: "The order of validation for username and display username.",
					skip: "prompt",
					isNestedObject: [{
						flag: "username-validation-order-username",
						question: "When should username validation occur?",
						description: "The order of username validation.",
						defaultValue: "pre-normalization",
						skip: "prompt",
						isSelectOptions: [{
							value: "pre-normalization",
							label: "Pre-normalization"
						}, {
							value: "post-normalization",
							label: "Post-normalization"
						}],
						argument: {
							index: 0,
							isProperty: "username",
							schema: z$1.enum(["pre-normalization", "post-normalization"]).optional()
						}
					}, {
						flag: "username-validation-order-display-username",
						question: "When should display username validation occur?",
						description: "The order of display username validation.",
						defaultValue: "pre-normalization",
						skip: "prompt",
						isSelectOptions: [{
							value: "pre-normalization",
							label: "Pre-normalization"
						}, {
							value: "post-normalization",
							label: "Post-normalization"
						}],
						argument: {
							index: 0,
							isProperty: "displayUsername",
							schema: z$1.enum(["pre-normalization", "post-normalization"]).optional()
						}
					}],
					argument: {
						index: 0,
						isProperty: "validationOrder"
					}
				}
			]
		},
		authClient: {
			function: "usernameClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "usernameClient" })],
				isNamedImport: false
			}]
		}
	},
	magicLink: {
		displayName: "Magic Link",
		auth: {
			function: "magicLink",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "magicLink" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "magic-link-expires-in",
					description: "Time in seconds until the magic link expires. Default is (60 * 5) 5 minutes",
					question: "[Magic Link] What is the expiration time for the magic link in seconds?",
					defaultValue: 300,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "expiresIn",
						schema: z$1.coerce.number().optional()
					}
				},
				{
					flag: "magic-link-send-magic-link",
					description: "Send magic link implementation.",
					question: "[Magic Link] What is the send magic link?",
					defaultValue: `async ({ email, url, token }, request) => {
	 // Send magic link to the user
	}`,
					skip: "prompt",
					isRequired: true,
					argument: {
						index: 0,
						isProperty: "sendMagicLink",
						schema: z$1.coerce.string()
					}
				},
				{
					flag: "magic-link-rate-limit",
					description: "Rate limit configuration. Default window is 60 seconds and max is 5 requests.",
					skip: "prompt",
					isNestedObject: [{
						flag: "magic-link-rate-limit-window",
						description: "Window in seconds. Default is 60 seconds.",
						question: "[Magic Link] What is the window in seconds?",
						defaultValue: 60,
						skip: "prompt",
						isNumber: true,
						argument: {
							index: 0,
							isProperty: "window",
							schema: z$1.coerce.number().optional()
						}
					}, {
						flag: "magic-link-rate-limit-max",
						description: "Max requests. Default is 5 requests.",
						question: "[Magic Link] What is the max requests?",
						defaultValue: 5,
						skip: "prompt",
						isNumber: true,
						argument: {
							index: 0,
							isProperty: "max",
							schema: z$1.coerce.number().optional()
						}
					}],
					argument: {
						index: 0,
						isProperty: "rateLimit"
					}
				},
				{
					flag: "magic-link-store-token",
					description: "This option allows you to configure how the token is stored in your database. Note: This will not affect the token that's sent, it will only affect the token stored in your database.",
					question: "[Magic Link] How would you like to store the token?",
					defaultValue: "plain",
					skip: "prompt",
					isSelectOptions: [{
						value: "plain",
						label: "Plain"
					}, {
						value: "hashed",
						label: "Hashed"
					}],
					argument: {
						index: 0,
						isProperty: "storeToken",
						schema: z$1.enum(["plain", "hashed"]).optional()
					}
				}
			]
		},
		authClient: {
			function: "magicLinkClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "magicLinkClient" })],
				isNamedImport: false
			}]
		}
	},
	emailOTP: {
		displayName: "Email OTP",
		auth: {
			function: "emailOTP",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "emailOTP" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "email-otp-send-verification-otp",
					description: "Function to send email verification",
					question: "[Email OTP] What is the send verification o t p?",
					defaultValue: `async ({ email, otp, type }, request) => {
 // Send email with OTP
}`,
					skip: "prompt",
					isRequired: true,
					argument: {
						index: 0,
						isProperty: "sendVerificationOTP",
						schema: z$1.coerce.string()
					}
				},
				{
					flag: "email-otp-otp-length",
					description: "Length of the OTP",
					question: "[Email OTP] What is the length of the OTP?",
					defaultValue: 6,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "otpLength",
						schema: z$1.coerce.number().optional()
					}
				},
				{
					flag: "email-otp-expires-in",
					description: "Expiry time of the OTP in seconds default is 5 minutes",
					question: "[Email OTP] What is the expiry time of the OTP in seconds?",
					defaultValue: 300,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "expiresIn",
						schema: z$1.coerce.number().optional()
					}
				},
				{
					flag: "email-otp-send-verification-on-sign-up",
					description: "Send email verification on sign-up",
					question: "[Email OTP] Would you like to send the OTP on sign-up?",
					defaultValue: false,
					skip: "prompt",
					isConfirmation: true,
					argument: {
						index: 0,
						isProperty: "sendVerificationOnSignUp",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "email-otp-disable-sign-up",
					description: "A boolean value that determines whether to prevent automatic sign-up when the user is not registered.",
					question: "[Email OTP] Would you like to disable sign-up?",
					defaultValue: false,
					skip: "prompt",
					isConfirmation: true,
					argument: {
						index: 0,
						isProperty: "disableSignUp",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "email-otp-allowed-attempts",
					description: "Allowed attempts for the OTP code",
					question: "[Email OTP] What is the allowed attempts for the OTP code?",
					defaultValue: 3,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "allowedAttempts",
						schema: z$1.coerce.number().optional()
					}
				},
				{
					flag: "email-otp-store-otp",
					description: "Store the OTP in your database in a secure way Note: This will not affect the OTP sent to the user, it will only affect the OTP stored in your database",
					question: "[Email OTP] How would you like to store the OTP code?",
					defaultValue: "plain",
					skip: "prompt",
					isSelectOptions: [
						{
							value: "plain",
							label: "Plain"
						},
						{
							value: "encrypted",
							label: "Encrypted"
						},
						{
							value: "hashed",
							label: "Hashed"
						}
					],
					argument: {
						index: 0,
						isProperty: "storeOTP",
						schema: z$1.enum([
							"plain",
							"encrypted",
							"hashed"
						]).optional()
					}
				},
				{
					flag: "email-otp-override-default-email-verification",
					description: "Override the default email verification to use email otp instead",
					question: "[Email OTP] Would you like to override the default email verification?",
					defaultValue: false,
					skip: "prompt",
					isConfirmation: true,
					argument: {
						index: 0,
						isProperty: "overrideDefaultEmailVerification",
						schema: z$1.coerce.boolean().optional()
					}
				}
			]
		},
		authClient: {
			function: "emailOTPClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "emailOTPClient" })],
				isNamedImport: false
			}]
		}
	},
	genericOAuth: {
		displayName: "Generic OAuth",
		auth: {
			function: "genericOAuth",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "genericOAuth" })],
				isNamedImport: false
			}]
		},
		authClient: {
			function: "genericOAuthClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "genericOAuthClient" })],
				isNamedImport: false
			}]
		}
	},
	anonymous: {
		displayName: "Anonymous",
		auth: {
			function: "anonymous",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "anonymous" })],
				isNamedImport: false
			}]
		},
		authClient: {
			function: "anonymousClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "anonymousClient" })],
				isNamedImport: false
			}]
		}
	},
	phoneNumber: {
		displayName: "Phone Number",
		auth: {
			function: "phoneNumber",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "phoneNumber" })],
				isNamedImport: false
			}]
		},
		authClient: {
			function: "phoneNumberClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "phoneNumberClient" })],
				isNamedImport: false
			}]
		}
	},
	passkey: {
		displayName: "Passkey",
		auth: {
			function: "passkey",
			imports: [{
				path: "better-auth/plugins/passkey",
				imports: [createImport({ name: "passkey" })],
				isNamedImport: false
			}]
		},
		authClient: {
			function: "passkeyClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "passkeyClient" })],
				isNamedImport: false
			}]
		}
	},
	oidc: {
		displayName: "OIDC",
		auth: {
			function: "oidc",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "oidc" })],
				isNamedImport: false
			}]
		},
		authClient: {
			function: "oidcClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "oidcClient" })],
				isNamedImport: false
			}]
		}
	},
	admin: {
		displayName: "Admin",
		auth: {
			function: "admin",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "admin" })],
				isNamedImport: false
			}],
			arguments: [{
				flag: "admin-default-role",
				question: "What is the default role for new users?",
				description: "The default role assigned to new users.",
				defaultValue: "user",
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "defaultRole",
					schema: z$1.coerce.string().optional()
				}
			}, {
				flag: "admin-roles",
				question: "What are the admin roles?",
				description: "Array of roles that are considered admin roles.",
				defaultValue: ["admin"],
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "adminRoles",
					schema: z$1.array(z$1.string()).optional()
				}
			}]
		},
		authClient: {
			function: "adminClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "adminClient" })],
				isNamedImport: false
			}]
		}
	},
	apiKey: {
		displayName: "API Key",
		auth: {
			function: "apiKey",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "apiKey" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "api-key-headers",
					question: "What header name should be used for API keys?",
					description: "The header name to check for API key.",
					defaultValue: "x-api-key",
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "apiKeyHeaders",
						schema: z$1.coerce.string().optional()
					}
				},
				{
					flag: "api-key-length",
					question: "What is the default length of API keys?",
					description: "The length of the API key. Longer is better.",
					defaultValue: 64,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "defaultKeyLength",
						schema: z$1.coerce.number().positive().optional()
					}
				},
				{
					flag: "api-key-disable-hashing",
					question: "Disable hashing of API keys?",
					description: "Disable hashing of the API key. Not recommended for security.",
					defaultValue: false,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "disableKeyHashing",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "api-key-enable-metadata",
					question: "Enable metadata for API keys?",
					description: "Whether to enable metadata for an API key.",
					defaultValue: false,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "enableMetadata",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "api-key-enable-session",
					question: "Enable session for API keys?",
					description: "An API Key can represent a valid session.",
					defaultValue: false,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "enableSessionForAPIKeys",
						schema: z$1.coerce.boolean().optional()
					}
				}
			]
		},
		authClient: {
			function: "apiKeyClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "apiKeyClient" })],
				isNamedImport: false
			}]
		}
	},
	bearer: {
		displayName: "Bearer",
		auth: {
			function: "bearer",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "bearer" })],
				isNamedImport: false
			}],
			arguments: [{
				flag: "bearer-require-signature",
				question: "Require signature for bearer tokens?",
				description: "If true, only signed tokens will be converted to session cookies.",
				defaultValue: false,
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "requireSignature",
					schema: z$1.coerce.boolean().optional()
				}
			}]
		},
		authClient: null
	},
	captcha: {
		displayName: "CAPTCHA",
		auth: {
			function: "captcha",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "captcha" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "captcha-provider",
					question: "Which CAPTCHA provider do you want to use?",
					description: "The CAPTCHA provider to use.",
					isSelectOptions: [
						{
							value: "google-recaptcha",
							label: "Google reCAPTCHA"
						},
						{
							value: "cloudflare-turnstile",
							label: "Cloudflare Turnstile"
						},
						{
							value: "hcaptcha",
							label: "hCaptcha"
						},
						{
							value: "captchafox",
							label: "CaptchaFox"
						}
					],
					argument: {
						index: 0,
						isProperty: "provider",
						schema: z$1.enum([
							"google-recaptcha",
							"cloudflare-turnstile",
							"hcaptcha",
							"captchafox"
						])
					}
				},
				{
					flag: "captcha-secret-key",
					question: "What is your CAPTCHA secret key?",
					description: "The secret key for the CAPTCHA provider.",
					argument: {
						index: 0,
						isProperty: "secretKey",
						schema: z$1.coerce.string()
					}
				},
				{
					flag: "captcha-site-key",
					question: "What is your CAPTCHA site key?",
					description: "The site key for the CAPTCHA provider (required for hCaptcha and CaptchaFox).",
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "siteKey",
						schema: z$1.coerce.string().optional()
					}
				},
				{
					flag: "captcha-min-score",
					question: "What is the minimum score for Google reCAPTCHA?",
					description: "The minimum score for Google reCAPTCHA v3 (0-1).",
					defaultValue: .5,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "minScore",
						schema: z$1.coerce.number().min(0).max(1).optional()
					}
				}
			]
		},
		authClient: null
	},
	customSession: {
		displayName: "Custom Session",
		auth: {
			function: "customSession",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "customSession" })],
				isNamedImport: false
			}],
			arguments: [{
				flag: "custom-session-mutate-list-device-sessions",
				question: "Should the list-device-sessions endpoint be mutated?",
				description: "Determine if the list-device-sessions endpoint should be mutated to the custom session data.",
				defaultValue: false,
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "shouldMutateListDeviceSessionsEndpoint",
					schema: z$1.coerce.boolean().optional()
				}
			}]
		},
		authClient: {
			function: "customSessionClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "customSessionClient" })],
				isNamedImport: false
			}]
		}
	},
	deviceAuthorization: {
		displayName: "Device Authorization",
		auth: {
			function: "deviceAuthorization",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "deviceAuthorization" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "device-auth-expires-in",
					question: "When should device codes expire?",
					description: "Time until the device code expires. Use formats like '30m', '5s', '1h'.",
					defaultValue: "30m",
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "expiresIn",
						schema: z$1.coerce.string().optional()
					}
				},
				{
					flag: "device-auth-interval",
					question: "What is the polling interval?",
					description: "Time between polling attempts. Use formats like '30m', '5s', '1h'.",
					defaultValue: "5s",
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "interval",
						schema: z$1.coerce.string().optional()
					}
				},
				{
					flag: "device-auth-device-code-length",
					question: "What is the length of the device code?",
					description: "Length of the device code to be generated.",
					defaultValue: 40,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "deviceCodeLength",
						schema: z$1.coerce.number().positive().optional()
					}
				},
				{
					flag: "device-auth-user-code-length",
					question: "What is the length of the user code?",
					description: "Length of the user code to be generated.",
					defaultValue: 8,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "userCodeLength",
						schema: z$1.coerce.number().positive().optional()
					}
				}
			]
		},
		authClient: {
			function: "deviceAuthorizationClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "deviceAuthorizationClient" })],
				isNamedImport: false
			}]
		}
	},
	haveIBeenPwned: {
		displayName: "Have I Been Pwned",
		auth: {
			function: "haveIBeenPwned",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "haveIBeenPwned" })],
				isNamedImport: false
			}],
			arguments: [{
				flag: "haveibeenpwned-custom-message",
				question: "What is the custom message for compromised passwords?",
				description: "Custom message to display when a password is compromised.",
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "customPasswordCompromisedMessage",
					schema: z$1.coerce.string().optional()
				}
			}]
		},
		authClient: null
	},
	jwt: {
		displayName: "JWT",
		auth: {
			function: "jwt",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "jwt" })],
				isNamedImport: false
			}],
			arguments: [{
				flag: "jwt-disable-setting-jwt-header",
				question: "Disable setting JWT header?",
				description: "If true, the JWT header will not be set in responses.",
				defaultValue: false,
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "disableSettingJwtHeader",
					schema: z$1.coerce.boolean().optional()
				}
			}]
		},
		authClient: {
			function: "jwtClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "jwtClient" })],
				isNamedImport: false
			}]
		}
	},
	lastLoginMethod: {
		displayName: "Last Login Method",
		auth: {
			function: "lastLoginMethod",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "lastLoginMethod" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "last-login-method-cookie-name",
					question: "What is the cookie name for last login method?",
					description: "Name of the cookie to store the last login method.",
					defaultValue: "better-auth.last_used_login_method",
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "cookieName",
						schema: z$1.coerce.string().optional()
					}
				},
				{
					flag: "last-login-method-max-age",
					question: "What is the cookie expiration time in seconds?",
					description: "Cookie expiration time in seconds.",
					defaultValue: 2592e3,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "maxAge",
						schema: z$1.coerce.number().positive().optional()
					}
				},
				{
					flag: "last-login-method-store-in-database",
					question: "Store the last login method in the database?",
					description: "Store the last login method in the database. This will create a new field in the user table.",
					defaultValue: false,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "storeInDatabase",
						schema: z$1.coerce.boolean().optional()
					}
				}
			]
		},
		authClient: {
			function: "lastLoginMethodClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "lastLoginMethodClient" })],
				isNamedImport: false
			}]
		}
	},
	mcp: {
		displayName: "MCP",
		auth: {
			function: "mcp",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "mcp" })],
				isNamedImport: false
			}],
			arguments: [{
				flag: "mcp-login-page",
				question: "What is the login page URL?",
				description: "The login page URL for MCP.",
				argument: {
					index: 0,
					isProperty: "loginPage",
					schema: z$1.coerce.string()
				}
			}, {
				flag: "mcp-resource",
				question: "What is the resource URL?",
				description: "The resource URL for MCP.",
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "resource",
					schema: z$1.coerce.string().optional()
				}
			}]
		},
		authClient: null
	},
	multiSession: {
		displayName: "Multi Session",
		auth: {
			function: "multiSession",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "multiSession" })],
				isNamedImport: false
			}],
			arguments: [{
				flag: "multi-session-maximum-sessions",
				question: "What is the maximum number of sessions a user can have?",
				description: "The maximum number of sessions a user can have at a time.",
				defaultValue: 5,
				skip: "prompt",
				isNumber: true,
				argument: {
					index: 0,
					isProperty: "maximumSessions",
					schema: z$1.coerce.number().positive().optional()
				}
			}]
		},
		authClient: {
			function: "multiSessionClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "multiSessionClient" })],
				isNamedImport: false
			}]
		}
	},
	oauthProxy: {
		displayName: "OAuth Proxy",
		auth: {
			function: "oAuthProxy",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "oAuthProxy" })],
				isNamedImport: false
			}],
			arguments: [{
				flag: "oauth-proxy-current-url",
				question: "What is the current URL of the application?",
				description: "The current URL of the application. The plugin will attempt to infer this from your environment.",
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "currentURL",
					schema: z$1.coerce.string().optional()
				}
			}, {
				flag: "oauth-proxy-production-url",
				question: "What is the production URL?",
				description: "If a request is in a production URL it won't be proxied.",
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "productionURL",
					schema: z$1.coerce.string().optional()
				}
			}]
		},
		authClient: null
	},
	oneTap: {
		displayName: "One Tap",
		auth: {
			function: "oneTap",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "oneTap" })],
				isNamedImport: false
			}],
			arguments: [{
				flag: "one-tap-disable-signup",
				question: "Disable the signup flow?",
				description: "Disable the signup flow.",
				defaultValue: false,
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "disableSignup",
					schema: z$1.coerce.boolean().optional()
				}
			}, {
				flag: "one-tap-client-id",
				question: "What is your Google Client ID?",
				description: "Google Client ID. If a client ID is provided in the social provider configuration, it will be used.",
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "clientId",
					schema: z$1.coerce.string().optional()
				}
			}]
		},
		authClient: {
			function: "oneTapClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "oneTapClient" })],
				isNamedImport: false
			}]
		}
	},
	oneTimeToken: {
		displayName: "One Time Token",
		auth: {
			function: "oneTimeToken",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "oneTimeToken" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "one-time-token-expires-in",
					question: "When should tokens expire (in minutes)?",
					description: "Expires in minutes.",
					defaultValue: 3,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "expiresIn",
						schema: z$1.coerce.number().positive().optional()
					}
				},
				{
					flag: "one-time-token-disable-client-request",
					question: "Disable client requests?",
					description: "Only allow server initiated requests.",
					defaultValue: false,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "disableClientRequest",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "one-time-token-store-token",
					question: "How should tokens be stored?",
					description: "Configure how the token is stored in your database.",
					defaultValue: "plain",
					skip: "prompt",
					isSelectOptions: [{
						value: "plain",
						label: "Plain text"
					}, {
						value: "hashed",
						label: "Hashed"
					}],
					argument: {
						index: 0,
						isProperty: "storeToken",
						schema: z$1.enum(["plain", "hashed"]).optional()
					}
				}
			]
		},
		authClient: {
			function: "oneTimeTokenClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "oneTimeTokenClient" })],
				isNamedImport: false
			}]
		}
	},
	openAPI: {
		displayName: "Open API",
		auth: {
			function: "openAPI",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "openAPI" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "open-api-path",
					question: "What is the path to the OpenAPI reference page?",
					description: "The path to the OpenAPI reference page. This will be appended to the base URL `/api/auth` path.",
					defaultValue: "/reference",
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "path",
						schema: z$1.coerce.string().optional()
					}
				},
				{
					flag: "open-api-disable-default-reference",
					question: "Disable the default reference page?",
					description: "Disable the default reference page that is generated by Scalar.",
					defaultValue: false,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "disableDefaultReference",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "open-api-theme",
					question: "What theme should be used for the OpenAPI reference page?",
					description: "Theme of the OpenAPI reference page.",
					defaultValue: "default",
					skip: "prompt",
					isSelectOptions: [
						{
							value: "alternate",
							label: "Alternate"
						},
						{
							value: "default",
							label: "Default"
						},
						{
							value: "moon",
							label: "Moon"
						},
						{
							value: "purple",
							label: "Purple"
						},
						{
							value: "solarized",
							label: "Solarized"
						},
						{
							value: "bluePlanet",
							label: "Blue Planet"
						},
						{
							value: "saturn",
							label: "Saturn"
						},
						{
							value: "kepler",
							label: "Kepler"
						},
						{
							value: "mars",
							label: "Mars"
						},
						{
							value: "deepSpace",
							label: "Deep Space"
						},
						{
							value: "laserwave",
							label: "Laserwave"
						},
						{
							value: "none",
							label: "None"
						}
					],
					argument: {
						index: 0,
						isProperty: "theme",
						schema: z$1.enum([
							"alternate",
							"default",
							"moon",
							"purple",
							"solarized",
							"bluePlanet",
							"saturn",
							"kepler",
							"mars",
							"deepSpace",
							"laserwave",
							"none"
						]).optional()
					}
				}
			]
		},
		authClient: null
	},
	organization: {
		displayName: "Organization",
		auth: {
			function: "organization",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "organization" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "organization-allow-user-to-create",
					question: "Allow users to create organizations?",
					description: "Configure whether new users are able to create new organizations.",
					defaultValue: true,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "allowUserToCreateOrganization",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "organization-creator-role",
					question: "What role should be assigned to the creator?",
					description: "The role that is assigned to the creator of the organization.",
					defaultValue: "owner",
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "creatorRole",
						schema: z$1.coerce.string().optional()
					}
				},
				{
					flag: "organization-membership-limit",
					question: "What is the maximum number of members allowed?",
					description: "The maximum number of members allowed in an organization.",
					defaultValue: 100,
					skip: "prompt",
					isNumber: true,
					argument: {
						index: 0,
						isProperty: "membershipLimit",
						schema: z$1.coerce.number().positive().optional()
					}
				}
			]
		},
		authClient: {
			function: "organizationClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "organizationClient" })],
				isNamedImport: false
			}]
		}
	},
	siwe: {
		displayName: "SIWE",
		auth: {
			function: "siwe",
			imports: [{
				path: "better-auth/plugins",
				imports: [createImport({ name: "siwe" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "siwe-domain",
					question: "What is the domain for SIWE?",
					description: "The domain for SIWE.",
					argument: {
						index: 0,
						isProperty: "domain",
						schema: z$1.coerce.string()
					}
				},
				{
					flag: "siwe-email-domain-name",
					question: "What is the email domain name?",
					description: "The email domain name for anonymous users.",
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "emailDomainName",
						schema: z$1.coerce.string().optional()
					}
				},
				{
					flag: "siwe-anonymous",
					question: "Allow anonymous users?",
					description: "Allow anonymous users.",
					defaultValue: true,
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "anonymous",
						schema: z$1.coerce.boolean().optional()
					}
				}
			]
		},
		authClient: {
			function: "siweClient",
			imports: [{
				path: "better-auth/client/plugins",
				imports: [createImport({ name: "siweClient" })],
				isNamedImport: false
			}]
		}
	},
	scim: {
		displayName: "SCIM",
		dependencies: ["@better-auth/scim"],
		auth: {
			function: "scim",
			imports: [{
				path: "@better-auth/scim",
				imports: [createImport({ name: "scim" })],
				isNamedImport: false
			}]
		},
		authClient: {
			function: "scimClient",
			imports: [{
				path: "@better-auth/scim/client",
				imports: [createImport({ name: "scimClient" })],
				isNamedImport: false
			}]
		}
	},
	sso: {
		displayName: "SSO",
		dependencies: ["@better-auth/sso"],
		auth: {
			function: "sso",
			imports: [{
				path: "@better-auth/sso",
				imports: [createImport({ name: "sso" })],
				isNamedImport: false
			}],
			arguments: [
				{
					flag: "sso-default-override-user-info",
					question: "Do you want to override the user info with the provider info?",
					description: "Override the user info with the provider info.",
					skip: "prompt",
					isConfirmation: true,
					defaultValue: false,
					argument: {
						index: 0,
						isProperty: "defaultOverrideUserInfo",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "sso-disable-implicit-sign-up",
					question: "Do you want to disable implicit sign up?",
					description: "Disable implicit sign up.",
					skip: "prompt",
					isConfirmation: true,
					defaultValue: false,
					argument: {
						index: 0,
						isProperty: "disableImplicitSignUp",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "sso-providers-limit",
					question: "What is the maximum number of SSO providers a user can register?",
					description: "The maximum number of SSO providers a user can register.",
					skip: "prompt",
					isNumber: true,
					defaultValue: 10,
					isRequired: false,
					argument: {
						index: 0,
						isProperty: "providersLimit",
						schema: z$1.coerce.number().int().positive().optional()
					}
				},
				{
					flag: "sso-trust-email-verified",
					question: "Do you want to trust the email verified flag from the provider?",
					description: "Trust the email verified flag from the provider.",
					skip: "prompt",
					isConfirmation: true,
					defaultValue: false,
					argument: {
						index: 0,
						isProperty: "trustEmailVerified",
						schema: z$1.coerce.boolean().optional()
					}
				},
				{
					flag: "sso-domain-verification",
					question: "Do you want to setup domain verification?",
					description: "Setup domain verification.",
					skip: "prompt",
					argument: {
						index: 0,
						isProperty: "domainVerification",
						schema: z$1.object({ enabled: z$1.coerce.boolean().optional() }).optional()
					},
					isNestedObject: [{
						flag: "sso-domain-verification-enabled",
						question: "Do you want to enable domain verification?",
						description: "Enable domain verification.",
						skip: "prompt",
						isConfirmation: true,
						defaultValue: false,
						argument: {
							index: 0,
							isProperty: "enabled",
							schema: z$1.coerce.boolean().optional()
						}
					}]
				}
			]
		},
		authClient: {
			function: "ssoClient",
			imports: [{
				path: "@better-auth/sso/client",
				imports: [createImport({ name: "ssoClient" })],
				isNamedImport: false
			}],
			arguments: [{
				flag: "sso-client-domain-verification",
				question: "Do you want to setup domain verification?",
				description: "Setup domain verification.",
				skip: "prompt",
				argument: {
					index: 0,
					isProperty: "domainVerification",
					schema: z$1.object({ enabled: z$1.coerce.boolean().optional() }).optional()
				},
				isNestedObject: [{
					flag: "sso-client-domain-verification-enabled",
					question: "Do you want to enable domain verification?",
					description: "Enable domain verification.",
					skip: "prompt",
					isConfirmation: true,
					defaultValue: false,
					argument: {
						index: 0,
						isProperty: "enabled",
						schema: z$1.coerce.boolean().optional()
					}
				}]
			}]
		}
	},
	stripe: {
		displayName: "Stripe",
		dependencies: ["stripe", "@better-auth/stripe"],
		auth: {
			function: "stripe",
			imports: [{
				path: "@better-auth/stripe",
				imports: [createImport({ name: "stripe" })],
				isNamedImport: false
			}]
		},
		authClient: {
			function: "stripeClient",
			imports: [],
			arguments: []
		}
	},
	i18n: {
		displayName: "I18n",
		dependencies: ["@better-auth/i18n"],
		auth: {
			function: "i18n",
			imports: [{
				path: "@better-auth/i18n",
				imports: [createImport({ name: "i18n" })],
				isNamedImport: false
			}],
			arguments: []
		},
		authClient: {
			function: "i18nClient",
			imports: [{
				path: "@better-auth/i18n/client",
				imports: [createImport({ name: "i18nClient" })],
				isNamedImport: false
			}]
		}
	}
};

//#endregion
//#region src/commands/init/utility/prompt.ts
const getFlagVariable = (flag) => {
	return flag.replace(/-([a-z])/g, (_, letter) => letter.toUpperCase());
};
const collectPromptableArgs = (args, options) => {
	if (!args) return [];
	const result = [];
	for (const arg of args) if (arg.isNestedObject && Array.isArray(arg.isNestedObject)) result.push(...collectPromptableArgs(arg.isNestedObject, options));
	else {
		const flagVar = getFlagVariable(arg.flag);
		const hasFlag = options[flagVar] !== void 0 && options[flagVar] !== "";
		if (arg.skip === "always" || arg.skip === "prompt") continue;
		if (arg.skip === "flag") result.push(arg);
		else if (!hasFlag) result.push(arg);
	}
	return result;
};
const toPromptQuestion = (arg) => {
	const base = {
		name: getFlagVariable(arg.flag),
		message: arg.question ?? arg.description ?? "",
		initial: arg.defaultValue
	};
	if (arg.isMultiselectOptions) return {
		...base,
		type: "multiselect",
		choices: arg.isMultiselectOptions.map((opt) => ({
			title: opt.label ?? String(opt.value),
			value: opt.value,
			description: opt.hint
		})),
		format: (v) => Array.isArray(v) ? v.join(", ") : v
	};
	if (arg.isSelectOptions) return {
		...base,
		type: "select",
		choices: arg.isSelectOptions.map((opt) => ({
			title: opt.label ?? String(opt.value),
			value: opt.value,
			description: opt.hint
		}))
	};
	if (arg.isConfirmation) return {
		...base,
		type: "confirm"
	};
	if (arg.isNumber) return {
		...base,
		type: "number",
		validate: (v) => arg.isRequired && (v == null || Number.isNaN(v)) ? "This field is required" : true
	};
	return {
		...base,
		type: "text",
		validate: (v) => {
			if (arg.isRequired && (!v || !v.trim())) return "This field is required";
			if (arg.argument.schema) {
				const parsed = arg.argument.schema.safeParse(arg.cliTransform ? arg.cliTransform(v) : v);
				return parsed.success ? true : parsed.error.message;
			}
			return true;
		}
	};
};
const getArgumentsPrompt = async (options, plugins, target) => {
	const opts = options;
	const allPromptableArgs = [];
	for (const plugin of plugins) if (target === "auth" && plugin.auth.arguments) allPromptableArgs.push(...collectPromptableArgs(plugin.auth.arguments, opts));
	else if (target === "authClient" && plugin.authClient && plugin.authClient.arguments) allPromptableArgs.push(...collectPromptableArgs(plugin.authClient.arguments, opts));
	let batchAnswers = {};
	if (allPromptableArgs.length > 0) batchAnswers = await prompts(allPromptableArgs.map(toPromptQuestion), { onCancel: () => {
		console.log("✋ Operation cancelled.");
		process.exit(0);
	} }) ?? {};
	return (arg) => {
		const flagVar = getFlagVariable(arg.flag);
		const hasFlag = opts[flagVar] !== void 0 && opts[flagVar] !== "";
		if (arg.skip === "always") return arg.isRequired && arg.defaultValue !== void 0 ? arg.defaultValue : void 0;
		if (arg.skip === "prompt") {
			if (hasFlag) {
				const val = opts[flagVar];
				return arg.cliTransform ? arg.cliTransform(val) : val;
			}
			return arg.isRequired && arg.defaultValue !== void 0 ? arg.defaultValue : void 0;
		}
		if (arg.skip === "flag") {
			if (batchAnswers[flagVar] !== void 0) {
				let val = batchAnswers[flagVar];
				if (arg.cliTransform) val = arg.cliTransform(val);
				return val;
			}
			return arg.isRequired && arg.defaultValue !== void 0 ? arg.defaultValue : void 0;
		}
		if (hasFlag) {
			const val = opts[flagVar];
			return arg.cliTransform ? arg.cliTransform(val) : val;
		}
		if (batchAnswers[flagVar] !== void 0) {
			let val = batchAnswers[flagVar];
			if (arg.cliTransform) val = arg.cliTransform(val);
			return val;
		}
		return arg.isRequired && arg.defaultValue !== void 0 ? arg.defaultValue : void 0;
	};
};

//#endregion
//#region src/commands/init/utility/plugin.ts
const getPluginConfigs = (plugins) => {
	return plugins.map((plugin) => {
		const pluginConfig = tempPluginsConfig[plugin];
		if (!pluginConfig) throw new Error(`Plugin ${plugin} not found`);
		return pluginConfig;
	});
};
/**
* Helper function to process nested arguments and build a nested object
*/
const processNestedArguments = async (nestedArguments, getArguments) => {
	const nestedObject = {};
	for (const nestedArg of nestedArguments) {
		let nestedValue;
		if (nestedArg.isNestedObject && Array.isArray(nestedArg.isNestedObject)) nestedValue = await processNestedArguments(nestedArg.isNestedObject, getArguments);
		else {
			let result = await getArguments(nestedArg);
			if (nestedArg.cliTransform) result = nestedArg.cliTransform(result);
			const schema = nestedArg.argument.schema?.safeParse(result) ?? {
				success: true,
				data: result
			};
			if (!schema.success) throw new Error(`Invalid nested argument: ${schema.error.message}`);
			nestedValue = schema.data;
		}
		if (nestedArg.argument.isProperty) {
			const propertyName = nestedArg.argument.isProperty;
			if (typeof nestedValue !== "undefined") if (nestedObject[propertyName] && typeof nestedObject[propertyName] === "object" && typeof nestedValue === "object" && nestedValue !== null && !(typeof nestedValue === "string" && nestedValue.includes("=>"))) nestedObject[propertyName] = {
				...nestedObject[propertyName],
				...nestedValue
			};
			else nestedObject[propertyName] = nestedValue;
		} else if (typeof nestedValue !== "undefined") throw new Error(`Nested argument must have isProperty set`);
	}
	return nestedObject;
};
/**
* Recursively clean objects by removing undefined values and setting empty nested objects to undefined
*/
const cleanNestedObjects = (value) => {
	if (typeof value === "undefined") return;
	if (typeof value === "string" && value.includes("=>")) return value;
	if (typeof value === "object" && value !== null && !Array.isArray(value)) {
		const cleaned = {};
		for (const [key, val] of Object.entries(value)) {
			const cleanedValue = cleanNestedObjects(val);
			if (typeof cleanedValue !== "undefined") cleaned[key] = cleanedValue;
		}
		if (Object.keys(cleaned).length === 0) return;
		return cleaned;
	}
	return value;
};
/**
* Process a single argument (handles both nested and regular arguments)
*/
const processArgument = async (argument, getArguments, functionName) => {
	let value;
	if (argument.isNestedObject && Array.isArray(argument.isNestedObject)) {
		value = await processNestedArguments(argument.isNestedObject, getArguments);
		if (argument.argument.schema) {
			const schema = argument.argument.schema.safeParse(value);
			if (!schema.success) throw new Error(`Invalid nested object for ${functionName}: ${schema.error.message}`);
			value = schema.data;
		}
	} else {
		let result = await getArguments(argument);
		if (argument.cliTransform) result = argument.cliTransform(result);
		const schema = argument.argument.schema?.safeParse(result) ?? {
			success: true,
			data: result
		};
		if (!schema.success) throw new Error(`Invalid argument for ${functionName} on flag "${argument.flag}": ${schema.error.message}`);
		value = schema.data;
	}
	return value;
};
/**
* Build argumentsCode map from arguments array
*/
const buildArgumentsCode = async (argumentOptions, getArguments, functionName) => {
	const argumentsCode = /* @__PURE__ */ new Map();
	if (!argumentOptions) return argumentsCode;
	for (const argument of argumentOptions) {
		const value = await processArgument(argument, getArguments, functionName);
		const index = argument.argument.index;
		if (argument.argument.isProperty) if (argumentsCode.has(index)) {
			const previous = argumentsCode.get(index) || {};
			if (typeof previous !== "object") throw new Error(`Argument at index ${index} is not an object`);
			argumentsCode.set(index, {
				...previous,
				[argument.argument.isProperty]: value
			});
		} else argumentsCode.set(index, { [argument.argument.isProperty]: value });
		else argumentsCode.set(index, value);
	}
	return argumentsCode;
};
/**
* Convert argumentsCode map to an array of string values
*/
const convertArgumentsCodeToStringArray = (argumentsCode) => {
	const hasFunctionString = (obj) => {
		for (const val of Object.values(obj)) {
			if (typeof val === "string" && val.includes("=>")) return true;
			if (typeof val === "object" && val !== null && !Array.isArray(val)) {
				if (hasFunctionString(val)) return true;
			}
		}
		return false;
	};
	const buildObjectString = (obj) => {
		return `{${Object.entries(obj).map(([key, val]) => {
			if (typeof val === "string" && val.includes("=>")) return `${key}: ${val}`;
			if (typeof val === "object" && val !== null && !Array.isArray(val)) return `${key}: ${buildObjectString(val)}`;
			return `${key}: ${JSON.stringify(val)}`;
		}).join(", ")}}`;
	};
	return Array.from(argumentsCode.values()).map((value) => {
		const cleaned = cleanNestedObjects(value);
		if (typeof cleaned === "undefined") return "undefined";
		if (typeof cleaned === "string" && cleaned.includes("=>")) return cleaned;
		if (typeof cleaned === "object" && cleaned !== null && !Array.isArray(cleaned)) {
			if (hasFunctionString(cleaned)) return buildObjectString(cleaned);
		}
		return JSON.stringify(cleaned);
	});
};
/**
* Remove trailing undefined values from args array
*/
const removeTrailingUndefined = (args) => {
	for (let i = args.length - 1; i >= 0; i--) {
		if (args[i] !== "undefined") break;
		args.pop();
	}
};
const getAuthPluginsCode = async ({ plugins, options = {}, installDependency }) => {
	if (!plugins || plugins.length === 0) return;
	const getArguments = await getArgumentsPrompt(options, plugins, "auth");
	const pluginsCode = [];
	for (const plugin of plugins) {
		const args = convertArgumentsCodeToStringArray(await buildArgumentsCode(plugin.auth.arguments, getArguments, plugin.auth.function));
		removeTrailingUndefined(args);
		pluginsCode.push(`${plugin.auth.function}(${args.join(", ")})`);
		const dependencies = new Set([...plugin.dependencies || [], ...plugin.auth.dependencies || []]);
		const devDependencies = new Set([...plugin.devDependencies || [], ...plugin.auth.devDependencies || []]);
		if (dependencies.size > 0) await installDependency([...dependencies]);
		if (devDependencies.size > 0) await installDependency([...devDependencies], "dev");
	}
	return (await formatCode(`[${pluginsCode.join(", ")}]`)).trim().slice(0, -1);
};
const getAuthClientPluginsCode = async ({ plugins, options = {}, installDependency }) => {
	if (!plugins || plugins.length === 0) return;
	const pluginsWithClient = plugins.filter((plugin) => plugin.authClient !== null);
	if (pluginsWithClient.length === 0) return;
	const getArguments = await getArgumentsPrompt(options, plugins, "authClient");
	const pluginsCode = [];
	for (const plugin of pluginsWithClient) {
		if (!plugin.authClient) continue;
		const args = convertArgumentsCodeToStringArray(await buildArgumentsCode(plugin.authClient.arguments, getArguments, plugin.authClient.function));
		removeTrailingUndefined(args);
		pluginsCode.push(`${plugin.authClient.function}(${args.join(", ")})`);
		const dependencies = new Set([...plugin.dependencies || [], ...plugin.authClient.dependencies || []]);
		const devDependencies = new Set([...plugin.devDependencies || [], ...plugin.authClient.devDependencies || []]);
		if (dependencies.size > 0) await installDependency([...dependencies]);
		if (devDependencies.size > 0) await installDependency([...devDependencies], "dev");
	}
	return (await formatCode(`[${pluginsCode.join(", ")}]`)).trim().slice(0, -1);
};

//#endregion
//#region src/commands/init/utility/auth-config.ts
const generateInnerAuthConfigCode = async ({ database, plugins, appName, baseURL, emailAndPassword, socialProviders, options, installDependency }) => {
	const code = {
		database: getDatabaseCode$1(database),
		appName: getAppNameCode(appName),
		baseURL: getBaseURLCode(baseURL),
		emailAndPassword: getEmailAndPasswordCode(emailAndPassword),
		socialProviders: getSocialProvidersCode(socialProviders),
		plugins: await getAuthPluginsCode({
			plugins,
			options,
			installDependency
		})
	};
	let stringCode = "";
	for (const key in code) {
		if (!code[key]) continue;
		stringCode += `${key}: ${code[key]},\n`;
	}
	return stringCode;
};
const getEmailAndPasswordCode = (enabled) => {
	if (!enabled) return void 0;
	return `{ enabled: true }`;
};
const getSocialProvidersCode = (providers) => {
	if (!providers || providers.length === 0) return void 0;
	return `{\n${providers.map((provider) => {
		const config = SOCIAL_PROVIDER_CONFIGS[provider];
		if (!config) {
			const providerUpper = provider.toUpperCase();
			return `		${provider}: {
			clientId: process.env.${providerUpper}_CLIENT_ID!,
			clientSecret: process.env.${providerUpper}_CLIENT_SECRET!,
		}`;
		}
		return `		${provider}: {\n${config.options.map((opt) => {
			return `			${opt.name}: process.env.${opt.envVar}!,`;
		}).join("\n")}\n		}`;
	}).join(",\n")}\n	}`;
};
const getAppNameCode = (appName) => {
	if (!appName) return;
	if (typeof appName !== "string") throw new Error("appName must be a string");
	return JSON.stringify(appName);
};
const getBaseURLCode = (baseURL) => {
	if (!baseURL) return;
	if (typeof baseURL !== "string") throw new Error("baseURL must be a string");
	let url;
	try {
		url = new URL(baseURL);
	} catch {
		throw new Error("baseURL must be a valid URL");
	}
	return JSON.stringify(url.toString());
};
const getDatabaseCode$1 = (database) => {
	if (!database) return void 0;
	return database.code({});
};

//#endregion
//#region src/commands/init/configs/databases.config.ts
const prismaCode = ({ provider, additionalOptions }) => {
	return `prismaAdapter(client, { provider: "${provider}", ${additionalOptions ? Object.entries(additionalOptions).map(([key, value]) => `${key}: ${value}`).join(", ") : ""} })`;
};
const drizzleCode = ({ provider, additionalOptions }) => {
	return `drizzleAdapter(db, { provider: "${provider}", schema, ${additionalOptions ? Object.entries(additionalOptions).map(([key, value]) => `${key}: ${value}`).join(", ") : ""} })`;
};
const kyselyCode = ({ provider, additionalOptions }) => {
	return `{dialect, type: "${provider}", ${additionalOptions ? Object.entries(additionalOptions).map(([key, value]) => `${key}: ${value}`).join(", ") : ""} }`;
};
const mongodbCode = ({ additionalOptions }) => {
	let optsString = "";
	if (additionalOptions) {
		optsString = ", {";
		optsString += Object.entries(additionalOptions).map(([key, value]) => `${key}: ${value}`).join(", ");
		optsString += "}";
	}
	return `mongodbAdapter(db${optsString})`;
};
const databasesConfig = [
	{
		adapter: "prisma-sqlite",
		imports: [{
			path: "better-auth/adapters/prisma",
			imports: [createImport({ name: "prismaAdapter" })],
			isNamedImport: false
		}, {
			path: "@prisma/client",
			imports: [createImport({ name: "PrismaClient" })],
			isNamedImport: false
		}],
		preCode: "const client = new PrismaClient();",
		code: ({ additionalOptions }) => {
			return prismaCode({
				provider: "sqlite",
				additionalOptions
			});
		},
		dependencies: ["@prisma/client"],
		devDependencies: ["prisma"]
	},
	{
		adapter: "prisma-mysql",
		imports: [{
			path: "better-auth/adapters/prisma",
			imports: [createImport({ name: "prismaAdapter" })],
			isNamedImport: false
		}, {
			path: "@prisma/client",
			imports: [createImport({ name: "PrismaClient" })],
			isNamedImport: false
		}],
		preCode: "const client = new PrismaClient();",
		code: ({ additionalOptions }) => {
			return prismaCode({
				provider: "mysql",
				additionalOptions
			});
		},
		dependencies: ["@prisma/client"],
		devDependencies: ["prisma"]
	},
	{
		adapter: "prisma-postgresql",
		imports: [{
			path: "better-auth/adapters/prisma",
			imports: [createImport({ name: "prismaAdapter" })],
			isNamedImport: false
		}, {
			path: "@prisma/client",
			imports: [createImport({ name: "PrismaClient" })],
			isNamedImport: false
		}],
		preCode: "const client = new PrismaClient();",
		code: ({ additionalOptions }) => {
			return prismaCode({
				provider: "postgresql",
				additionalOptions
			});
		},
		dependencies: ["@prisma/client"],
		devDependencies: ["prisma"]
	},
	{
		adapter: "drizzle-sqlite-better-sqlite3",
		imports: [
			{
				path: "better-auth/adapters/drizzle",
				imports: [createImport({ name: "drizzleAdapter" })],
				isNamedImport: false
			},
			{
				path: "drizzle-orm/better-sqlite3",
				imports: [createImport({ name: "drizzle" })],
				isNamedImport: false
			},
			{
				path: "better-sqlite3",
				imports: createImport({ name: "Database" }),
				isNamedImport: true
			},
			{
				path: "./auth-schema",
				imports: createImport({
					name: "*",
					alias: "schema"
				}),
				isNamedImport: true
			}
		],
		preCode: `const db = drizzle(new Database("database.sqlite"), { schema });`,
		code({ additionalOptions }) {
			return drizzleCode({
				provider: "sqlite",
				additionalOptions
			});
		},
		dependencies: ["drizzle-orm", "better-sqlite3"],
		devDependencies: ["@types/better-sqlite3"]
	},
	{
		adapter: "drizzle-sqlite-bun",
		imports: [
			{
				path: "better-auth/adapters/drizzle",
				imports: [createImport({ name: "drizzleAdapter" })],
				isNamedImport: false
			},
			{
				path: "drizzle-orm/bun-sqlite",
				imports: [createImport({ name: "drizzle" })],
				isNamedImport: false
			},
			{
				path: "bun:sqlite",
				imports: [createImport({ name: "Database" })],
				isNamedImport: false
			},
			{
				path: "./auth-schema",
				imports: createImport({
					name: "*",
					alias: "schema"
				}),
				isNamedImport: true
			}
		],
		preCode: `const db = drizzle({ client: new Database('sqlite.db'), schema });`,
		code({ additionalOptions }) {
			return drizzleCode({
				provider: "sqlite",
				additionalOptions
			});
		},
		dependencies: ["drizzle-orm", "bun"],
		devDependencies: ["@types/bun"]
	},
	{
		adapter: "drizzle-postgresql",
		imports: [
			{
				path: "better-auth/adapters/drizzle",
				imports: [createImport({ name: "drizzleAdapter" })],
				isNamedImport: false
			},
			{
				path: "drizzle-orm/node-postgres",
				imports: [createImport({ name: "drizzle" })],
				isNamedImport: false
			},
			{
				path: "pg",
				imports: [createImport({ name: "Pool" })],
				isNamedImport: false
			},
			{
				path: "./auth-schema",
				imports: createImport({
					name: "*",
					alias: "schema"
				}),
				isNamedImport: true
			}
		],
		preCode: `const db = drizzle(new Pool({ connectionString: process.env.DATABASE_URL }), { schema });`,
		code({ additionalOptions }) {
			return drizzleCode({
				provider: "pg",
				additionalOptions
			});
		},
		dependencies: ["drizzle-orm", "pg"],
		devDependencies: ["@types/pg"]
	},
	{
		adapter: "drizzle-mysql",
		imports: [
			{
				path: "better-auth/adapters/drizzle",
				imports: [createImport({ name: "drizzleAdapter" })],
				isNamedImport: false
			},
			{
				path: "drizzle-orm/mysql2",
				imports: [createImport({ name: "drizzle" })],
				isNamedImport: false
			},
			{
				path: "mysql2/promise",
				imports: [createImport({ name: "createPool" })],
				isNamedImport: false
			},
			{
				path: "./auth-schema",
				imports: createImport({
					name: "*",
					alias: "schema"
				}),
				isNamedImport: true
			}
		],
		preCode: `const db = drizzle(createPool(process.env.DATABASE_URL!), { schema, mode: "default" });`,
		code({ additionalOptions }) {
			return drizzleCode({
				provider: "mysql",
				additionalOptions
			});
		},
		dependencies: ["drizzle-orm", "mysql2"]
	},
	{
		adapter: "sqlite-better-sqlite3",
		imports: [{
			path: "better-sqlite3",
			imports: createImport({ name: "Database" }),
			isNamedImport: true
		}],
		preCode: `const database = new Database("auth.db")`,
		code({ additionalOptions }) {
			return `database`;
		},
		dependencies: ["better-sqlite3"],
		devDependencies: ["@types/better-sqlite3"]
	},
	{
		adapter: "sqlite-bun",
		imports: [{
			path: "bun:sqlite",
			imports: [createImport({ name: "Database" })],
			isNamedImport: false
		}],
		preCode: `const database = new Database("auth.db")`,
		code({ additionalOptions }) {
			return `database`;
		},
		dependencies: []
	},
	{
		adapter: "sqlite-node",
		imports: [{
			path: "node:sqlite",
			imports: [createImport({ name: "DatabaseSync" })],
			isNamedImport: false
		}],
		preCode: `const database = new DatabaseSync("auth.db")`,
		code({ additionalOptions }) {
			return `database`;
		},
		dependencies: []
	},
	{
		adapter: "mysql",
		imports: [{
			path: "mysql2/promise",
			imports: [createImport({ name: "createPool" })],
			isNamedImport: false
		}],
		preCode: `const dialect = createPool({ host: "localhost", user: "root", password: "password", database: "database", timezone: "Z" })`,
		code({ additionalOptions }) {
			return kyselyCode({
				provider: "mysql",
				additionalOptions
			});
		},
		dependencies: ["mysql2"]
	},
	{
		adapter: "postgresql",
		imports: [{
			path: "pg",
			imports: [createImport({ name: "Pool" })],
			isNamedImport: false
		}],
		preCode: `const dialect = new Pool({ connectionString: "postgresql://postgres:password@localhost:5432/database" })`,
		code({ additionalOptions }) {
			return kyselyCode({
				provider: "postgresql",
				additionalOptions
			});
		},
		dependencies: ["pg"],
		devDependencies: ["@types/pg"]
	},
	{
		adapter: "mssql",
		imports: [
			{
				path: "kysely",
				imports: [createImport({ name: "MssqlDialect" })],
				isNamedImport: false
			},
			{
				path: "tedious",
				imports: createImport({
					name: "*",
					alias: "Tedious"
				}),
				isNamedImport: true
			},
			{
				path: "tarn",
				imports: createImport({
					name: "*",
					alias: "Tarn"
				}),
				isNamedImport: true
			}
		],
		preCode: `const dialect = new MssqlDialect({
                    tarn: {
                        ...Tarn,
                        options: {
                        min: 0,
                        max: 10,
                        },
                    },
                    tedious: {
                        ...Tedious,
                        connectionFactory: () => new Tedious.Connection({
                        authentication: {
                            options: {
                            password: 'password',
                            userName: 'username',
                            },
                            type: 'default',
                        },
                        options: {
                            database: 'some_db',
                            port: 1433,
                            trustServerCertificate: true,
                        },
                        server: 'localhost',
                        }),
						TYPES: {
								...Tedious.TYPES,
								DateTime: Tedious.TYPES.DateTime2,
							},
                    },
                    })`,
		code({ additionalOptions }) {
			return kyselyCode({
				provider: "mssql",
				additionalOptions
			});
		},
		dependencies: [
			"kysely",
			"tedious",
			"tarn"
		]
	},
	{
		adapter: "mongodb",
		imports: [{
			path: "better-auth/adapters/mongodb",
			imports: [createImport({ name: "mongodbAdapter" })],
			isNamedImport: false
		}, {
			path: "mongodb",
			imports: [createImport({ name: "MongoClient" })],
			isNamedImport: false
		}],
		preCode: `const client = new MongoClient(process.env.DATABASE_URL!);\nconst db = client.db();`,
		code({ additionalOptions }) {
			return mongodbCode({ additionalOptions });
		},
		dependencies: ["mongodb"]
	}
];

//#endregion
//#region src/commands/init/utility/database.ts
const getDatabaseCode = (adapter) => {
	if (!adapter) return null;
	return databasesConfig.find((database) => database.adapter === adapter);
};
/**
* Extract ORM name from adapter string
* Examples:
* - "prisma-sqlite" -> "prisma"
* - "drizzle-postgresql" -> "drizzle"
* - "drizzle-sqlite-better-sqlite3" -> "drizzle"
* - "sqlite-better-sqlite3" -> "kysely"
* - "sqlite-bun" -> "kysely"
* - "mongodb" -> "mongodb"
*/
const getORMFromAdapter = (adapter) => {
	if (adapter.includes("-")) {
		const parts = adapter.split("-");
		if (parts[0] === "sqlite" && parts.length > 1) return "kysely";
		return parts[0];
	}
	if ([
		"mysql",
		"postgresql",
		"mssql"
	].includes(adapter)) return "kysely";
	return adapter;
};
/**
* Check if an adapter is a kysely dialect
*/
const isKyselyDialect = (adapter) => {
	return adapter.startsWith("sqlite-") || [
		"mysql",
		"postgresql",
		"mssql"
	].includes(adapter);
};
/**
* Check if an adapter should be returned directly without dialect selection
* (Kysely dialects and MongoDB don't have sub-dialects)
*/
const isDirectAdapter = (adapter) => {
	return isKyselyDialect(adapter) || adapter === "mongodb";
};
/**
* Format adapter name for display
*/
const formatAdapterLabel = (adapter) => {
	if (adapter === "sqlite-better-sqlite3") return "SQLite (better-sqlite3)";
	if (adapter === "sqlite-bun") return "SQLite (bun)";
	if (adapter === "sqlite-node") return "SQLite (node:sqlite)";
	if (adapter === "drizzle-sqlite-better-sqlite3") return "SQLite (better-sqlite3)";
	if (adapter === "drizzle-sqlite-bun") return "SQLite (bun)";
	if (adapter === "drizzle-sqlite-node") return "SQLite (node:sqlite)";
	return adapter.charAt(0).toUpperCase() + adapter.slice(1);
};
/**
* Get all unique ORMs from the database config
* SQLite variants are grouped under a single "SQLite" option
*/
const getAvailableORMs = () => {
	const options = [];
	const seenORMs = /* @__PURE__ */ new Set();
	for (const db of databasesConfig) {
		const dbORM = getORMFromAdapter(db.adapter);
		if (db.adapter.startsWith("sqlite-")) {
			if (!seenORMs.has("sqlite")) {
				seenORMs.add("sqlite");
				options.push({
					value: "sqlite",
					label: "SQLite"
				});
			}
		} else if (dbORM === "kysely" || dbORM === "mongodb") options.push({
			value: db.adapter,
			label: formatAdapterLabel(db.adapter),
			adapter: db.adapter
		});
		else if (!seenORMs.has(dbORM)) {
			seenORMs.add(dbORM);
			options.push({
				value: dbORM,
				label: dbORM.charAt(0).toUpperCase() + dbORM.slice(1)
			});
		}
	}
	const sortOrder = [
		"sqlite",
		"postgresql",
		"mysql",
		"drizzle",
		"prisma",
		"mongodb",
		"mssql"
	];
	return options.sort((a, b) => {
		const aIndex = sortOrder.indexOf(a.value);
		const bIndex = sortOrder.indexOf(b.value);
		if (aIndex !== -1 && bIndex !== -1) return aIndex - bIndex;
		if (aIndex !== -1) return -1;
		if (bIndex !== -1) return 1;
		return a.value.localeCompare(b.value);
	});
};
/**
* Get available dialects for a specific ORM
*/
const getDialectsForORM = (orm) => {
	const dialects = [];
	for (const db of databasesConfig) if (getORMFromAdapter(db.adapter) === orm) {
		let label;
		if (db.adapter.includes("-")) {
			const parts = db.adapter.split("-");
			if (orm === "drizzle" && parts[1] === "sqlite") label = formatAdapterLabel(db.adapter);
			else {
				const dialectName = parts.slice(1).join("-");
				label = dialectName.charAt(0).toUpperCase() + dialectName.slice(1);
			}
		} else label = formatAdapterLabel(db.adapter);
		dialects.push({
			value: db.adapter,
			label,
			adapter: db.adapter
		});
	}
	return dialects.sort((a, b) => a.value.localeCompare(b.value));
};

//#endregion
//#region src/commands/init/generate-auth.ts
const generateAuthConfigCode = async ({ plugins: pluginsConfig, database: databaseConfig, appName, baseURL, emailAndPassword, socialProviders, installDependency, options }) => {
	const database = getDatabaseCode(databaseConfig);
	const plugins = getPluginConfigs(pluginsConfig);
	const imports = [
		{
			imports: [createImport({ name: "betterAuth" })],
			path: "better-auth",
			isNamedImport: false
		},
		...Object.values(plugins).map(({ auth }) => auth.imports).flat(),
		...database?.imports ?? []
	];
	const authConfigCode = await generateInnerAuthConfigCode({
		plugins,
		database,
		appName,
		baseURL,
		emailAndPassword,
		socialProviders,
		options,
		installDependency
	});
	const segmentedCode = {
		imports: await getImportString(imports),
		exports: "",
		preAuthConfig: database?.preCode ?? "",
		authConfig: authConfigCode,
		postAuthConfig: ""
	};
	return await formatCode([
		segmentedCode.imports,
		``,
		segmentedCode.preAuthConfig,
		``,
		`export const auth = betterAuth({`,
		segmentedCode.authConfig,
		`});`,
		``,
		segmentedCode.postAuthConfig,
		``,
		segmentedCode.exports
	].join("\n"));
};

//#endregion
//#region src/commands/init/utility/auth-client-config.ts
const generateInnerAuthClientConfigCode = async ({ plugins, options, installDependency }) => {
	const code = { plugins: await getAuthClientPluginsCode({
		plugins,
		options,
		installDependency
	}) };
	let stringCode = "";
	for (const key in code) {
		if (!code[key]) continue;
		stringCode += `${key}: ${code[key]},\n`;
	}
	return stringCode;
};

//#endregion
//#region src/commands/init/generate-auth-client.ts
const generateAuthClientConfigCode = async ({ plugins: pluginsConfig, database: databaseConfig, framework, options, installDependency }) => {
	const plugins = getPluginConfigs(pluginsConfig);
	const imports = [...!framework.authClient ? [{
		imports: [createImport({ name: "createAuthClient" })],
		path: "better-auth/client",
		isNamedImport: false
	}] : [{
		imports: [createImport({ name: "createAuthClient" })],
		path: framework.authClient.importPath,
		isNamedImport: false
	}], ...Object.values(plugins).map(({ authClient }) => !authClient ? [] : authClient.imports).flat()];
	const authClientCode = await generateInnerAuthClientConfigCode({
		plugins,
		options,
		installDependency
	});
	const segmentedCode = {
		imports: await getImportString(imports),
		exports: "",
		preAuthConfig: "",
		authConfig: authClientCode ? `{${authClientCode}}` : "",
		postAuthConfig: ""
	};
	return await formatCode([
		segmentedCode.imports,
		``,
		segmentedCode.preAuthConfig,
		``,
		`export const authClient = createAuthClient(`,
		segmentedCode.authConfig,
		`);`,
		``,
		segmentedCode.postAuthConfig,
		``,
		segmentedCode.exports
	].join("\n"));
};

//#endregion
//#region src/commands/init/utility/env.ts
const getEnvFiles = async (cwd) => {
	return (await fs$1.readdir(cwd, "utf-8")).filter((file) => file.startsWith(".env") && file !== ".env.example").map((file) => path.join(cwd, file));
};
const parseEnvFiles = async (envFiles) => {
	const result = /* @__PURE__ */ new Map();
	for (const file of envFiles) {
		const existingVars = (await fs$1.readFile(file, "utf-8")).split("\n").filter((line) => line.trim()).map((x) => x.split("=")[0]).filter((x) => x && x.trim()).filter((x) => !x?.includes(" ")).filter((x) => !x?.startsWith("#"));
		result.set(file, existingVars);
	}
	return result;
};
const updateEnvFiles = async (envFiles, envs) => {
	for (const file of envFiles) {
		const lines = (await fs$1.readFile(file, "utf-8")).split("\n");
		lines.push(...envs);
		await fs$1.writeFile(file, lines.join("\n"), "utf-8");
	}
};
/**
* Gets the missing env variables in the env files
*
* @param envFiles - The list of env files to check
* @param envVar - The env variable to check
* @returns The list of env files that are missing the env variable
*/
const getMissingEnvVars = async (envFiles, envVar) => {
	const missingVarInFiles = [];
	for (const [file, existingVars] of envFiles) if (Array.isArray(envVar)) {
		const missingVars = envVar.filter((v) => !existingVars.includes(v));
		if (missingVars.length > 0) missingVarInFiles.push({
			file,
			var: missingVars
		});
	} else if (typeof envVar === "string" && !existingVars.includes(envVar)) missingVarInFiles.push({
		file,
		var: [envVar]
	});
	return missingVarInFiles;
};
const createEnvFile = async (cwd, envVariables) => {
	const envFile = path.join(cwd, ".env");
	await fs$1.writeFile(envFile, envVariables.join("\n"), "utf-8");
};

//#endregion
//#region src/commands/init/utility/framework.ts
async function detectFramework(cwd, packageJson) {
	for (const strategy of [packageJsonStrategy, fileStrategy]) {
		const result = await strategy({
			cwd,
			packageJson
		});
		if (result !== null) return result;
	}
	return null;
}
const packageJsonStrategy = ({ packageJson }) => {
	for (const framework of FRAMEWORKS) if (hasDependency(packageJson, framework.dependency)) return framework;
	return null;
};
const fileStrategy = ({ cwd }) => {
	const cwdFiles = readdirSync(cwd);
	for (const framework of FRAMEWORKS) {
		if (!framework.configPaths?.length) continue;
		for (const configPath of framework.configPaths) if (cwdFiles.includes(configPath)) return framework;
	}
	return null;
};

//#endregion
//#region src/commands/init/index.ts
const confirm = async (options) => {
	return (await prompts({
		type: "confirm",
		name: "value",
		message: options.message,
		initial: options.initial ?? true
	}))?.value ?? null;
};
const select = async (options) => {
	return (await prompts({
		type: "select",
		name: "value",
		message: options.message,
		choices: options.options.map((opt) => ({
			title: opt.label,
			value: opt.value
		})),
		initial: options.initialValue ? options.options.findIndex((opt) => opt.value === options.initialValue) : void 0
	}))?.value ?? null;
};
const multiselect = async (options) => {
	return (await prompts({
		type: "multiselect",
		name: "value",
		message: options.message,
		choices: options.options.map((opt) => ({
			title: opt.label,
			value: opt.value
		})),
		instructions: false
	}))?.value ?? null;
};
const isCancel = (value) => {
	return value === null || value === void 0;
};
const cancel = (message) => {
	console.log(message);
	process.exit(0);
};
const log$1 = {
	info: (message) => console.log(message),
	success: (message) => console.log(chalk.green(message)),
	error: (message) => console.error(chalk.red(message))
};
/**
* Extract database provider from database adapter string
*/
const getDatabaseProvider = (database) => {
	if (database.startsWith("drizzle-")) {
		if (database.includes("postgresql")) return "pg";
		if (database.includes("mysql")) return "mysql";
		if (database.includes("sqlite")) return "sqlite";
	}
	if (database.startsWith("prisma-")) {
		if (database.includes("postgresql")) return "postgresql";
		if (database.includes("mysql")) return "mysql";
		if (database.includes("sqlite")) return "sqlite";
	}
	return null;
};
/**
* Create a minimal BetterAuthOptions config for schema generation
*/
const createMinimalConfig = (plugins, baseURL) => {
	return {
		secret: "temp-secret-for-schema-generation",
		baseURL,
		plugins: plugins.map((pluginKey) => {
			if (!tempPluginsConfig[pluginKey]) return null;
			return null;
		}).filter(Boolean)
	};
};
/**
* Create a mock adapter object for schema generation
*/
const createMockAdapter = (database, provider) => {
	const isDrizzle = database.startsWith("drizzle-");
	const isPrisma = database.startsWith("prisma-");
	return {
		id: isDrizzle ? "drizzle" : isPrisma ? "prisma" : "unknown",
		options: { provider: provider === "pg" ? "pg" : provider }
	};
};
/**
* Generate the correct import path for the auth file in route handlers
*/
const generateAuthImportPath = async (cwd, authFilePath, routeHandlerPath, framework) => {
	const absoluteAuthPath = path.resolve(cwd, authFilePath);
	const resolvedRouteHandlerPath = path.resolve(cwd, routeHandlerPath);
	const routeHandlerDir = path.dirname(resolvedRouteHandlerPath);
	if (framework?.id === "sveltekit") {
		const normalizedPath = path.relative(cwd, absoluteAuthPath).replace(/\\/g, "/");
		if (normalizedPath.startsWith("src/lib/") || normalizedPath === "src/lib") {
			const pathWithoutExt = normalizedPath.slice(8).replace(/\.(ts|js|tsx|jsx)$/, "");
			return pathWithoutExt ? `$lib/${pathWithoutExt}` : "$lib/auth";
		}
	}
	if (framework?.id === "hono") {
		let relativePath = path.relative(routeHandlerDir, absoluteAuthPath);
		relativePath = relativePath.replace(/\.(ts|js|tsx|jsx)$/, "");
		if (!relativePath.startsWith(".")) relativePath = `./${relativePath}`;
		return relativePath.replace(/\\/g, "/");
	}
	const tsconfigPath = path.join(cwd, "tsconfig.json");
	const { data: tsconfigContent } = await tryCatch(fs$1.readFile(tsconfigPath, "utf-8"));
	let aliasPrefix = null;
	let aliasBasePath = null;
	if (tsconfigContent) try {
		const cleanedContent = tsconfigContent.replace(/\/\*[\s\S]*?\*\/|\/\/.*/g, "");
		const compilerOptions = JSON.parse(cleanedContent)?.compilerOptions;
		const paths = compilerOptions?.paths;
		const baseUrl = compilerOptions?.baseUrl;
		if (paths) {
			for (const [alias, targets] of Object.entries(paths)) if (typeof alias === "string" && alias.endsWith("/*") && Array.isArray(targets) && targets.length > 0) {
				const target = targets[0];
				if (target.endsWith("/*")) {
					aliasPrefix = alias.slice(0, -2);
					let basePath = target.slice(0, -2);
					if (baseUrl && baseUrl !== ".") basePath = path.join(baseUrl, basePath);
					aliasBasePath = basePath;
					break;
				}
			}
		}
	} catch (_e) {}
	const relativeAuthPath = path.relative(cwd, absoluteAuthPath);
	if (aliasPrefix && aliasBasePath !== null) {
		const normalizedBasePath = path.normalize(aliasBasePath);
		console.log(chalk.dim(`    normalizedBasePath: ${normalizedBasePath}`));
		if (normalizedBasePath === "." || normalizedBasePath === "" || normalizedBasePath === "./") {
			const pathWithoutExt = relativeAuthPath.replace(/\.(ts|js|tsx|jsx)$/, "");
			const result = `${aliasPrefix}/${pathWithoutExt}`.replace(/\\/g, "/");
			console.log(chalk.dim(`    Using alias, returning: ${result}`));
			return result;
		}
		const normalizedRelativePath = relativeAuthPath.replace(/\\/g, "/");
		const normalizedBasePathForward = normalizedBasePath.replace(/\\/g, "/");
		if (normalizedRelativePath === normalizedBasePathForward || normalizedRelativePath.startsWith(normalizedBasePathForward + "/")) {
			let pathAfterBase;
			if (normalizedRelativePath === normalizedBasePathForward) pathAfterBase = "";
			else pathAfterBase = normalizedRelativePath.slice(normalizedBasePathForward.length + 1);
			const pathWithoutExt = pathAfterBase.replace(/\.(ts|js|tsx|jsx)$/, "");
			return pathWithoutExt ? `${aliasPrefix}/${pathWithoutExt}` : aliasPrefix;
		}
	}
	let relativePath = path.relative(routeHandlerDir, absoluteAuthPath);
	relativePath = relativePath.replace(/\.(ts|js|tsx|jsx)$/, "");
	if (!relativePath.startsWith(".")) relativePath = `./${relativePath}`;
	return relativePath.replace(/\\/g, "/");
};
async function initAction(opts) {
	const options = initActionOptionsSchema.parse(opts);
	const cwd = options.cwd;
	let packageJson = null;
	try {
		packageJson = await getPackageInfo(cwd);
	} catch {}
	if (typeof packageJson !== "object" || packageJson === null) {
		const pm = options.packageManager || "npm";
		const initCommand = pm === "bun" ? "bun init" : pm === "yarn" ? "yarn init" : `${pm} init`;
		console.error(chalk.red(`\nThis appears to be an empty project. No package.json found.\n`));
		console.error(chalk.yellow(`Please initialize a new project first by running:\n\n  ${chalk.bold(initCommand)}\n`));
		process.exit(1);
	}
	let currentStep = 0;
	const nextStep = async (text) => {
		currentStep++;
		console.log(chalk.white(`\n${currentStep}. ${text}`));
	};
	const additionalSteps = [];
	console.log("\n" + [
		`   ██  ████`,
		`   ████  ██  ${chalk.bold(`Better Auth CLI`)} ${chalk.dim(`(${cliVersion})`)}`,
		`   ██  ████  ${chalk.gray("Welcome to the Better Auth CLI! Let's get you set up.")}`
	].join("\n"));
	const { pm, pmString: _pmString } = await (async () => {
		if (options.packageManager) {
			const [pm, version] = [options.packageManager, null];
			return {
				pm,
				pmString: getPkgManagerStr({
					packageManager: pm,
					version
				})
			};
		}
		const { packageManager, version } = await detectPackageManager(cwd, packageJson);
		return {
			pm: packageManager,
			pmString: getPkgManagerStr({
				packageManager,
				version
			})
		};
	})();
	const depsToInstall = /* @__PURE__ */ new Map();
	const filesToWrite = [];
	await (async () => {
		if (await hasDependency(packageJson, "better-auth")) return;
		await nextStep("Install Better Auth");
		const shouldInstallBetterAuth = await confirm({
			message: `Would you like to install better-auth using ${chalk.bold(pm)}?`,
			initial: true
		});
		if (isCancel(shouldInstallBetterAuth)) {
			cancel("✋ Operation cancelled.");
			process.exit(0);
		}
		if (shouldInstallBetterAuth) depsToInstall.set("better-auth", { prod: true });
	})();
	let envFiles = /* @__PURE__ */ new Map();
	await (async () => {
		envFiles = await parseEnvFiles(await getEnvFiles(cwd));
		if (envFiles.size === 0) {
			await nextStep("Set Environment Variables");
			const shouldCreateEnv = await confirm({ message: `Would you like to set environment variables?` });
			if (isCancel(shouldCreateEnv)) {
				cancel("✋ Operation cancelled.");
				process.exit(0);
			}
			if (shouldCreateEnv) {
				const { providedSecret } = await prompts({
					type: "text",
					name: "providedSecret",
					message: `Better Auth secret (used for encryption, hashing, and signing). ${chalk.dim("(Press Enter to auto generate)")}`
				});
				if (isCancel(providedSecret)) {
					cancel("✋ Operation cancelled.");
					process.exit(0);
				}
				const { providedURL } = await prompts({
					type: "text",
					name: "providedURL",
					message: `Better Auth Base URL (your auth server URL):`,
					initial: "http://localhost:3000"
				});
				if (isCancel(providedURL)) {
					cancel("✋ Operation cancelled.");
					process.exit(0);
				}
				const envs = [`BETTER_AUTH_SECRET="${providedSecret || generateSecretHash$1()}"`, `BETTER_AUTH_URL="${providedURL}"`];
				envFiles.set(".env", envs);
				filesToWrite.push(() => createEnvFile(cwd, envs));
			}
			return;
		}
		const missingEnvVars = await getMissingEnvVars(envFiles, ["BETTER_AUTH_SECRET", "BETTER_AUTH_URL"]);
		if (!missingEnvVars.length) return;
		await nextStep("Set Environment Variables");
		if (missingEnvVars.length === 1) {
			const { file, var: missingVars } = missingEnvVars[0];
			const confirmed = await confirm({ message: `Add required environment variables to ${chalk.bold(file.split("/").pop())}? (${missingVars.map((v) => chalk.cyan(v)).join(", ")})` });
			if (isCancel(confirmed)) {
				cancel("✋ Operation cancelled.");
				process.exit(0);
			}
			if (confirmed) {
				const envs = [];
				for (const v of missingVars) if (v === "BETTER_AUTH_SECRET") {
					const { providedSecret } = await prompts({
						type: "text",
						name: "providedSecret",
						message: `Better Auth secret (used for encryption, hashing, and signing). ${chalk.dim("(Press Enter to auto generate)")}`
					});
					if (isCancel(providedSecret)) {
						cancel("✋ Operation cancelled.");
						process.exit(0);
					}
					envs.push(`BETTER_AUTH_SECRET="${providedSecret || generateSecretHash$1()}"`);
				} else if (v === "BETTER_AUTH_URL") {
					const { providedURL } = await prompts({
						type: "text",
						name: "providedURL",
						message: `Better Auth base URL (your auth server URL):`,
						initial: "http://localhost:3000"
					});
					if (isCancel(providedURL)) {
						cancel("✋ Operation cancelled.");
						process.exit(0);
					}
					envs.push(`BETTER_AUTH_URL="${providedURL}"`);
				}
				envFiles.set(file, envs);
				filesToWrite.push(() => updateEnvFiles([file], envs));
			}
			return;
		}
		const filesToUpdate = await multiselect({
			message: `Add required environment variables to the following files?`,
			options: missingEnvVars.map((x) => ({
				value: x.file,
				label: `${chalk.bold(x.file)}: ${x.var.map((v) => chalk.cyan(v)).join(", ")}`
			}))
		});
		if (isCancel(filesToUpdate)) {
			cancel("✋ Operation cancelled.");
			process.exit(0);
		}
		if (filesToUpdate) {
			const secretHash = generateSecretHash$1();
			for (const file of filesToUpdate) {
				const envs = missingEnvVars.find((x) => x.file === file).var.map((v) => {
					if (v === "BETTER_AUTH_SECRET") return `BETTER_AUTH_SECRET="${secretHash}"`;
					if (v === "BETTER_AUTH_URL") return "BETTER_AUTH_URL=\"http://localhost:3000\"";
					return `${v}=${v}`;
				});
				filesToWrite.push(() => updateEnvFiles([file], envs));
			}
			return;
		}
	})();
	const detectedFramework = await detectFramework(cwd, packageJson);
	let framework = detectedFramework || FRAMEWORKS.find((f) => f.id === "next");
	const frameworkWasDetected = !!detectedFramework;
	if (framework.id === "next" && framework.routeHandler) {
		const { data: rootFiles } = await tryCatch(fs$1.readdir(cwd, "utf-8"));
		const hasAppDir = rootFiles?.some((file) => file === "app");
		const hasPagesDir = rootFiles?.some((file) => file === "pages");
		const hasSrcDir = rootFiles?.some((file) => file === "src");
		let routeHandlerPath = "app/api/auth/[...all]/route.ts";
		if (hasSrcDir) {
			const { data: srcFiles } = await tryCatch(fs$1.readdir(path.join(cwd, "src"), "utf-8"));
			const hasSrcApp = srcFiles?.some((file) => file === "app");
			if (srcFiles?.some((file) => file === "pages")) routeHandlerPath = "src/pages/api/auth/[...all].ts";
			else if (hasSrcApp) routeHandlerPath = "src/app/api/auth/[...all]/route.ts";
		} else if (hasPagesDir) routeHandlerPath = "pages/api/auth/[...all].ts";
		else if (hasAppDir) routeHandlerPath = "app/api/auth/[...all]/route.ts";
		framework = {
			...framework,
			routeHandler: {
				...framework.routeHandler,
				path: routeHandlerPath
			}
		};
	}
	let authConfigFilePath = null;
	const hasAuthConfigAlready = await (async () => {
		for (const path_ of possibleAuthConfigPaths) {
			const fullPath = path.join(cwd, path_);
			const { error } = await tryCatch(fs$1.access(fullPath, fs$1.constants.F_OK));
			if (!error) {
				authConfigFilePath = fullPath;
				return true;
			}
		}
		return false;
	})();
	if (!hasAuthConfigAlready) {
		await nextStep("Create A Better Auth Instance");
		const { data: allFiles, error } = await tryCatch(fs$1.readdir(cwd, "utf-8"));
		if (error) {
			log$1.error(`Failed to read directory: ${error.message}`);
			process.exit(1);
		}
		const hasSrc = allFiles.some((node) => node === "src");
		let defaultAuthConfigPath;
		if (hasSrc) defaultAuthConfigPath = path.join(cwd, "src", "lib", "auth.ts");
		else defaultAuthConfigPath = path.join(cwd, "lib", "auth.ts");
		const { filePath } = await prompts({
			type: "text",
			name: "filePath",
			message: `Where would you like to create the auth instance?`,
			initial: path.relative(cwd, defaultAuthConfigPath)
		});
		if (isCancel(filePath)) {
			cancel("✋ Operation cancelled.");
			process.exit(0);
		}
		const cleanPath = filePath.startsWith("/") ? filePath.slice(1) : filePath;
		const absoluteFilePath = path.isAbsolute(cleanPath) ? cleanPath : path.join(cwd, cleanPath);
		authConfigFilePath = absoluteFilePath;
		const boilerplateCode = `import { betterAuth } from "better-auth";

export const auth = betterAuth({
	// Configuration will be added here
});
`;
		filesToWrite.push(async () => {
			const { error: mkdirError } = await tryCatch(fs$1.mkdir(path.dirname(absoluteFilePath), { recursive: true }));
			if (mkdirError) {
				const error = `Failed to create auth directory at ${path.dirname(absoluteFilePath)}: ${mkdirError.message}`;
				log$1.error(error);
				process.exit(1);
			}
			const { error: writeFileError } = await tryCatch(fs$1.writeFile(absoluteFilePath, boilerplateCode, "utf-8"));
			if (writeFileError) {
				const error = `Failed to write auth file at ${absoluteFilePath}: ${writeFileError.message}`;
				log$1.error(error);
				process.exit(1);
			}
		});
	}
	let databaseChoice = null;
	let database = null;
	let shouldGenerateSchema = false;
	let shouldRunMigration = false;
	await (async () => {
		await nextStep("Configure Database");
		const dbChoice = await select({
			message: `Would you like to configure a database?`,
			options: [
				{
					value: "yes",
					label: "Yes - Configure a database"
				},
				{
					value: "stateless",
					label: "Stateless - Skip database (stateless mode)"
				},
				{
					value: "skip",
					label: "Skip - Don't setup database now"
				}
			]
		});
		if (isCancel(dbChoice)) {
			cancel("✋ Operation cancelled.");
			process.exit(0);
		}
		databaseChoice = dbChoice || null;
		if (databaseChoice === "yes") {
			const selectedOption = await select({
				message: `Select the database you want to use:`,
				options: getAvailableORMs().map((opt) => ({
					value: opt.adapter || opt.value,
					label: opt.label
				}))
			});
			if (isCancel(selectedOption)) {
				cancel("✋ Operation cancelled.");
				process.exit(0);
			}
			if (selectedOption === "sqlite") {
				const sqliteOptions = [];
				sqliteOptions.push({
					value: "sqlite-better-sqlite3",
					label: "better-sqlite3"
				});
				if (pm === "bun") sqliteOptions.push({
					value: "sqlite-bun",
					label: "Bun SQLite"
				});
				else sqliteOptions.push({
					value: "sqlite-node",
					label: "Node SQLite"
				});
				const sqliteVariants = await select({
					message: `Select SQLite driver:`,
					options: sqliteOptions
				});
				if (isCancel(sqliteVariants)) {
					cancel("✋ Operation cancelled.");
					process.exit(0);
				}
				database = sqliteVariants;
			} else if (isDirectAdapter(selectedOption)) database = selectedOption;
			else {
				const selectedDialect = await select({
					message: `Select the database dialect:`,
					options: getDialectsForORM(selectedOption).map((d) => ({
						value: d.adapter,
						label: d.label
					}))
				});
				if (isCancel(selectedDialect)) {
					cancel("✋ Operation cancelled.");
					process.exit(0);
				}
				database = selectedDialect;
			}
		}
		if (database) {
			const databaseConfig = getDatabaseCode(database);
			if (databaseConfig && databaseConfig.dependencies.length > 0) {
				const { shouldInstallDeps } = await prompts({
					type: "confirm",
					name: "shouldInstallDeps",
					message: `Would you like to install the following dependencies: ${[...new Set([...databaseConfig.dependencies, ...databaseConfig.devDependencies || []])].map((x) => chalk.cyan(x)).join(", ")}?`,
					initial: true
				});
				if (isCancel(shouldInstallDeps)) {
					cancel("✋ Operation cancelled.");
					process.exit(0);
				}
				if (shouldInstallDeps) {
					for (const dep of databaseConfig.dependencies) depsToInstall.set(dep, {
						...depsToInstall.get(dep) || {},
						prod: true
					});
					for (const dep of databaseConfig.devDependencies || []) depsToInstall.set(dep, {
						...depsToInstall.get(dep) || {},
						dev: true
					});
				}
			}
			const dbString = String(database);
			const isDrizzle = dbString.startsWith("drizzle-");
			const isPrisma = dbString.startsWith("prisma-");
			const isKysely = isKyselyDialect(dbString);
			const isMongoDB = dbString === "mongodb";
			if (isDrizzle || isPrisma) {
				const response = await prompts({
					type: "confirm",
					name: "shouldGenerate",
					message: `Would you like to generate the database schema?`,
					initial: true
				});
				if (isCancel(response.shouldGenerate)) {
					cancel("✋ Operation cancelled.");
					process.exit(0);
				}
				shouldGenerateSchema = response.shouldGenerate || false;
				if (shouldGenerateSchema) console.log(chalk.dim(`\n  Schema will be generated after auth configuration is complete.\n`));
			}
			if (isKysely) {
				const response = await prompts({
					type: "confirm",
					name: "shouldMigrate",
					message: `Would you like to run database migration?`,
					initial: true
				});
				if (isCancel(response.shouldMigrate)) {
					cancel("✋ Operation cancelled.");
					process.exit(0);
				}
				shouldRunMigration = response.shouldMigrate || false;
				if (shouldRunMigration) console.log(chalk.dim(`\n  Migration will run after auth configuration is complete.\n`));
			}
			if (isMongoDB) console.log(chalk.dim(`\n  MongoDB adapter will automatically create collections as needed.\n`));
		}
	})();
	let emailAndPassword = false;
	if (databaseChoice && databaseChoice !== "stateless") {
		await nextStep("Configure Email & Password");
		const confirmed = await confirm({
			message: `Would you like to enable email & password authentication?`,
			initial: true
		});
		if (isCancel(confirmed)) {
			cancel("✋ Operation cancelled.");
			process.exit(0);
		}
		emailAndPassword = confirmed || false;
	}
	let selectedSocialProviders = [];
	await (async () => {
		await nextStep("Configure Social Providers");
		const shouldSetupSocial = await confirm({
			message: `Would you like to setup social providers?`,
			initial: true
		});
		if (isCancel(shouldSetupSocial)) {
			cancel("✋ Operation cancelled.");
			process.exit(0);
		}
		if (shouldSetupSocial) {
			const providers = await multiselect({
				message: `Select the social providers you want to enable:`,
				options: SOCIAL_PROVIDERS.map((provider) => ({
					value: provider,
					label: provider.charAt(0).toUpperCase() + provider.slice(1)
				}))
			});
			if (isCancel(providers)) {
				cancel("✋ Operation cancelled.");
				process.exit(0);
			}
			selectedSocialProviders = providers || [];
		}
	})();
	if (selectedSocialProviders.length > 0) await (async () => {
		if (envFiles.size === 0) return;
		const socialProviderEnvVars = selectedSocialProviders.flatMap((provider) => {
			const config = SOCIAL_PROVIDER_CONFIGS[provider];
			if (!config) {
				const providerUpper = provider.toUpperCase();
				return [`${providerUpper}_CLIENT_ID`, `${providerUpper}_CLIENT_SECRET`];
			}
			return config.options.map((opt) => opt.envVar);
		});
		const missingSocialEnvVars = await getMissingEnvVars(envFiles, socialProviderEnvVars);
		if (missingSocialEnvVars.length > 0 && envFiles.size > 0) {
			const firstEnvFile = [...envFiles.keys()].sort((a, b) => path.basename(a).length - path.basename(b).length)[0];
			const envVarsToAdd = missingSocialEnvVars.filter((x) => x.file === firstEnvFile).flatMap((x) => x.var.map((v) => `${v}=""`));
			if (envVarsToAdd.length > 0) {
				const resolvedPath = path.isAbsolute(firstEnvFile) ? firstEnvFile : path.join(cwd, firstEnvFile);
				filesToWrite.push(() => updateEnvFiles([resolvedPath], envVarsToAdd));
			}
		}
	})();
	const plugins = await (async () => {
		return [];
	})();
	await (async () => {
		if (!authConfigFilePath) return;
		const authConfigCode = await generateAuthConfigCode({
			plugins,
			database,
			framework,
			baseURL: "http://localhost:3000",
			emailAndPassword,
			socialProviders: selectedSocialProviders,
			options,
			installDependency: (d, type) => {
				const dependencies = Array.isArray(d) ? d : [d];
				for (const dep of dependencies) depsToInstall.set(dep, {
					...depsToInstall.get(dep) || {},
					[type || "prod"]: true
				});
			}
		});
		filesToWrite.push(async () => {
			if (!authConfigFilePath) return;
			const { error: writeFileError } = await tryCatch(fs$1.writeFile(authConfigFilePath, authConfigCode, "utf-8"));
			if (writeFileError) {
				const error = `Failed to write auth file at ${authConfigFilePath}: ${writeFileError.message}`;
				log$1.error(error);
				process.exit(1);
			}
		});
	})();
	await (async () => {
		if (hasAuthConfigAlready) return;
		if (!database) return;
		const dbString = String(database);
		if (dbString === "mongodb") return;
		const isDrizzle = dbString.startsWith("drizzle-");
		const isPrisma = dbString.startsWith("prisma-");
		if (isKyselyDialect(dbString) && shouldRunMigration) {
			additionalSteps.push(async () => {
				await nextStep("Migrate Database");
				const s = yoctoSpinner({
					text: "Running database migration...",
					color: "white"
				});
				s.start();
				await new Promise((resolve, reject) => {
					exec(`npx auth migrate`, { cwd }, (error, stdout, stderr) => {
						if (error) {
							s.stop();
							log$1.error(`Failed to run migration: ${error.message}`);
							if (stderr) log$1.error(stderr);
							reject(error);
							return;
						}
						s.success("Database migration completed successfully!");
						if (stdout) console.log(stdout);
						resolve();
					});
				});
			});
			return;
		}
		if (!isDrizzle && !isPrisma) return;
		if (!shouldGenerateSchema) return;
		const provider = getDatabaseProvider(dbString);
		if (!provider) {
			log$1.error(`Unable to determine database provider for ${database}`);
			return;
		}
		const s = yoctoSpinner({
			text: `Generating database schema...`,
			color: "white"
		});
		s.start();
		try {
			const config = createMinimalConfig(plugins, "http://localhost:3000");
			const adapter = createMockAdapter(database, provider);
			let schemaResult;
			let outputPath;
			if (isDrizzle) {
				if (!authConfigFilePath) throw new Error("Auth config file path is required for Drizzle schema generation");
				const resolvedAuthConfigPath = path.isAbsolute(authConfigFilePath) ? authConfigFilePath : path.join(cwd, authConfigFilePath);
				const authConfigDir = path.dirname(resolvedAuthConfigPath);
				const fullOutputPath = path.join(authConfigDir, "auth-schema.ts");
				outputPath = path.relative(cwd, fullOutputPath);
				schemaResult = await generateDrizzleSchema({
					adapter,
					options: config,
					file: outputPath
				});
			} else if (isPrisma) {
				outputPath = "prisma/schema.prisma";
				schemaResult = await generatePrismaSchema({
					adapter,
					options: config,
					file: outputPath
				});
			} else throw new Error(`Unsupported database type: ${dbString}`);
			if (!schemaResult.code) {
				s.stop();
				log$1.info("Schema is already up to date.");
				return;
			}
			const fullOutputPath = path.isAbsolute(outputPath) ? outputPath : path.join(cwd, outputPath);
			if (await fs$1.access(fullOutputPath).then(() => true).catch(() => false) && schemaResult.overwrite) {
				s.stop();
				const shouldOverwrite = await confirm({
					message: `The file ${chalk.yellow(outputPath)} already exists. Do you want to overwrite it?`,
					initial: false
				});
				if (isCancel(shouldOverwrite) || !shouldOverwrite) {
					log$1.info("Schema generation cancelled.");
					return;
				}
				s.start();
			}
			filesToWrite.push(async () => {
				if (!schemaResult.code) return;
				const outputDir = path.dirname(fullOutputPath);
				await fs$1.mkdir(outputDir, { recursive: true });
				await fs$1.writeFile(fullOutputPath, schemaResult.code, "utf-8");
			});
			s.success(`Schema generated successfully at ${chalk.yellow(outputPath)}!`);
		} catch (error) {
			s.stop();
			log$1.error(`Failed to generate schema: ${error instanceof Error ? error.message : String(error)}`);
			process.exit(1);
		}
	})();
	await (async () => {
		if (!frameworkWasDetected) return;
		if (!framework.routeHandler) return;
		if (!authConfigFilePath) return;
		const { routeHandler } = framework;
		const fullPath = path.resolve(cwd, routeHandler.path);
		const { error } = await tryCatch(fs$1.access(fullPath, fs$1.constants.F_OK));
		if (!error) return;
		await nextStep("Generate Route Handler");
		const { filePath } = await prompts({
			type: "text",
			name: "filePath",
			message: `Enter the path to the route handler file:`,
			initial: path.relative(cwd, fullPath)
		});
		if (isCancel(filePath)) {
			cancel("✋ Operation cancelled.");
			process.exit(0);
		}
		const cleanHandlerPath = filePath.startsWith("/") ? filePath.slice(1) : filePath;
		const absoluteHandlerPath = path.isAbsolute(cleanHandlerPath) ? cleanHandlerPath : path.join(cwd, cleanHandlerPath);
		const authImportPath = await generateAuthImportPath(cwd, authConfigFilePath, absoluteHandlerPath, framework);
		let updatedCode = routeHandler.code;
		for (const pattern of [
			/from\s+["']@\/[^"']+["']/,
			/from\s+["']~\/[^"']+["']/,
			/from\s+["']\$lib\/[^"']+["']/,
			/from\s+["']\.\/[^"']+["']/,
			/from\s+["']\.\.\/[^"']+["']/
		]) {
			const newCode = updatedCode.replace(pattern, `from "${authImportPath}"`);
			if (newCode !== updatedCode) {
				updatedCode = newCode;
				break;
			}
		}
		filesToWrite.push(async () => {
			const { error: mkdirError } = await tryCatch(fs$1.mkdir(path.dirname(absoluteHandlerPath), { recursive: true }));
			if (mkdirError) {
				const error = `Failed to create directory at ${path.dirname(absoluteHandlerPath)}: ${mkdirError.message}`;
				log$1.error(error);
				process.exit(1);
			}
			const { error: writeFileError } = await tryCatch(fs$1.writeFile(absoluteHandlerPath, updatedCode, "utf-8"));
			if (writeFileError) {
				const error = `Failed to write file at ${absoluteHandlerPath}: ${writeFileError.message}`;
				log$1.error(error);
				process.exit(1);
			}
		});
	})();
	await (async () => {
		if (await (async () => {
			for (const path_ of possibleClientConfigPaths) {
				const fullPath = path.join(cwd, path_);
				const { error } = await tryCatch(fs$1.access(fullPath, fs$1.constants.F_OK));
				if (!error) return true;
			}
			return false;
		})()) return;
		await nextStep("Generate Auth Client Configuration");
		console.log(chalk.dim("Note: If you have a separated client-server project architecture, you may want to skip generating the auth client file here and create it in your client project instead."));
		const shouldGenerateAuthClient = await confirm({
			message: `Would you like to generate the auth client configuration file?`,
			initial: true
		});
		if (isCancel(shouldGenerateAuthClient)) {
			cancel("✋ Operation cancelled.");
			process.exit(0);
		}
		if (!shouldGenerateAuthClient) return;
		const authClientCode = await generateAuthClientConfigCode({
			plugins,
			database,
			framework,
			baseURL: "http://localhost:3000",
			options,
			installDependency: (d, type) => {
				const dependencies = Array.isArray(d) ? d : [d];
				for (const dep of dependencies) depsToInstall.set(dep, {
					...depsToInstall.get(dep) || {},
					[type || "prod"]: true
				});
			}
		});
		const { data: allFiles, error } = await tryCatch(fs$1.readdir(cwd, "utf-8"));
		if (error) {
			log$1.error(`Failed to read directory: ${error.message}`);
			process.exit(1);
		}
		const hasSrc = allFiles.some((node) => node === "src");
		let defaultAuthClientPath;
		if (hasSrc) defaultAuthClientPath = path.join(cwd, "src", "lib", "auth-client.ts");
		else defaultAuthClientPath = path.join(cwd, "lib", "auth-client.ts");
		const { filePath } = await prompts({
			type: "text",
			name: "filePath",
			message: `Enter the path to the auth-client.ts file:`,
			initial: path.relative(cwd, defaultAuthClientPath)
		});
		if (isCancel(filePath)) {
			cancel("✋ Operation cancelled.");
			process.exit(0);
		}
		const cleanPath = filePath.startsWith("/") ? filePath.slice(1) : filePath;
		const absoluteClientFilePath = path.isAbsolute(cleanPath) ? cleanPath : path.join(cwd, cleanPath);
		filesToWrite.push(async () => {
			const { error: mkdirError } = await tryCatch(fs$1.mkdir(path.dirname(absoluteClientFilePath), { recursive: true }));
			if (mkdirError) {
				const error = `Failed to create auth client directory at ${path.dirname(absoluteClientFilePath)}: ${mkdirError.message}`;
				log$1.error(error);
				process.exit(1);
			}
			const { error: writeFileError } = await tryCatch(fs$1.writeFile(absoluteClientFilePath, authClientCode, "utf-8"));
			if (writeFileError) {
				const error = `Failed to write auth client file at ${absoluteClientFilePath}: ${writeFileError.message}`;
				log$1.error(error);
				process.exit(1);
			}
		});
	})();
	await (async () => {
		if (filesToWrite.length === 0) return;
		await nextStep("Generate Files");
		const s = yoctoSpinner({
			text: "Generating files...",
			color: "white"
		});
		s.start();
		for (const exec of filesToWrite) await exec();
		s.success("Files generated successfully!");
	})();
	await (async () => {
		if (depsToInstall.size === 0) return;
		await nextStep("Install Dependencies");
		const s = yoctoSpinner({
			text: "Installing dependencies...",
			color: "white"
		});
		s.start();
		const deps = {
			prod: /* @__PURE__ */ new Set(),
			dev: /* @__PURE__ */ new Set()
		};
		for (const [dep, cfg] of depsToInstall) {
			if (cfg.prod) deps.prod.add(dep);
			if (cfg.dev) deps.dev.add(dep);
		}
		for (const [type, dependencies] of Object.entries(deps)) await installDependencies({
			cwd,
			dependencies: [...dependencies],
			packageManager: pm,
			type
		});
		s.success("Dependencies installed successfully!");
	})();
	for (const step of additionalSteps) await step();
	const connectResponse = await prompts({
		type: "confirm",
		name: "connect",
		message: "Would you like to connect your app to Better Auth infrastructure?",
		initial: true
	});
	if (connectResponse.connect === void 0) {
		console.log(chalk.yellow("\n✖ ") + "Setup cancelled before connecting to Better Auth infrastructure.\n");
		return;
	}
	if (connectResponse.connect === void 0) {
		console.log(chalk.yellow("\n✖ ") + "Setup cancelled before connecting to Better Auth infrastructure.\n");
		return;
	}
	if (connectResponse.connect) {
		await open("https://beta.better-auth.com/onboarding");
		console.log(chalk.cyan("\n→ ") + "Opening Better Auth onboarding in your browser...\n");
	}
	console.log(chalk.green(`\n✔ `) + chalk.bold("Success! ") + "Project setup complete.\n");
	const logs = [];
	let nextStepNum = 1;
	if (databaseChoice === "yes" && database) {
		logs.push(`  ${nextStepNum}. Set up your database with necessary environment variables`);
		nextStepNum++;
		const dbString = String(database);
		const isDrizzle = dbString.startsWith("drizzle-");
		const isPrisma = dbString.startsWith("prisma-");
		const isKysely = isKyselyDialect(dbString);
		if ((isDrizzle || isPrisma || isKysely) && !shouldRunMigration) {
			let command;
			if (isDrizzle) command = "npx drizzle-kit push";
			else if (isPrisma) command = "npx prisma migrate dev";
			else command = "npx auth migrate";
			logs.push(`  ${nextStepNum}. Run ${chalk.cyan(command)} to apply schema`);
			nextStepNum++;
		}
	}
	if (!frameworkWasDetected) {
		logs.push(`  ${nextStepNum}. Mount the auth handler`);
		logs.push(`     Use ${chalk.cyan("auth.handler")} with a Web API compatible request object\n     Default route: ${chalk.cyan("\"/api/auth\"")} (configurable via ${chalk.cyan("basePath")})`);
		nextStepNum++;
	}
	if (selectedSocialProviders.length > 0) {
		const providerList = selectedSocialProviders.map((provider) => {
			const config = SOCIAL_PROVIDER_CONFIGS[provider];
			if (!config) {
				const providerUpper = provider.toUpperCase();
				return `\n     - ${chalk.cyan(`${providerUpper}_CLIENT_ID`)} and ${chalk.cyan(`${providerUpper}_CLIENT_SECRET`)}`;
			}
			return `\n     - ${config.options.map((opt) => chalk.cyan(opt.envVar)).join(" and ")}`;
		}).join("");
		logs.push(`  ${nextStepNum}. Add social provider credentials to .env:${providerList}`);
		nextStepNum++;
	}
	if (logs.length > 0) {
		console.log(chalk.bold("Next steps:"));
		console.log(logs.join("\n"));
	}
}
const initBuilder = new Command("init").option("-c, --cwd <cwd>", "The working directory.", process.cwd()).option("--config <config>", "The path to the auth configuration file. defaults to the first `auth.ts` file found.").option("--package-manager <package-manager>", "The package manager to use. defaults to the package manager found in the current working directory.");
/**
* Track used flags to ensure uniqueness
*/
const usedFlags = /* @__PURE__ */ new Set();
/**
* Recursively process arguments and nested objects to add CLI options
* Each flag is unique and not compound (no parent prefix)
*/
const processArguments = (args, pluginDisplayName) => {
	if (!args) return;
	for (const argument of args) if (argument.isNestedObject && Array.isArray(argument.isNestedObject)) processArguments(argument.isNestedObject, pluginDisplayName);
	else {
		const flag = argument.flag;
		if (usedFlags.has(flag)) {
			console.warn(`Warning: Flag "${flag}" is already used. Skipping duplicate.`);
			continue;
		}
		usedFlags.add(flag);
		initBuilder.option(`--${flag} <${flag}>`, `[${pluginDisplayName}] ${argument.description}`);
		pluginArgumentOptionsSchema[getFlagVariable(flag)] = z.coerce.string().optional();
	}
};
const pluginArgumentOptionsSchema = {};
for (const plugin of Object.values(tempPluginsConfig)) {
	if (plugin.auth.arguments) processArguments(plugin.auth.arguments, plugin.displayName);
	if (plugin.authClient && plugin.authClient.arguments) processArguments(plugin.authClient.arguments, plugin.displayName);
}
const init = initBuilder.action(initAction);
const initActionOptionsSchema = z.object({
	cwd: z.string().transform((val) => path.resolve(val)),
	config: z.string().optional(),
	packageManager: z.enum(PACKAGE_MANAGER).optional(),
	...pluginArgumentOptionsSchema
});

//#endregion
//#region src/commands/login.ts
async function loginAction() {
	try {
		await spawnCommand("npx @better-auth/cli@latest login");
	} catch (error) {
		log.error(error.message || "An unknown error occurred");
		process.exit(1);
	}
	process.exit(0);
}
const login = new Command("login").description("Login to Better Auth Infrastructure").action(loginAction);
async function logoutAction() {
	try {
		await spawnCommand("npx @better-auth/cli@latest logout");
	} catch (error) {
		log.error(error.message || "An unknown error occurred");
		process.exit(1);
	}
	process.exit(0);
}
const logout = new Command("logout").description("Logout from Better Auth Infrastructure").action(logoutAction);

//#endregion
//#region src/commands/mcp.ts
const REMOTE_MCP_URL = "https://mcp.inkeep.com/better-auth/mcp";
async function mcpAction(options) {
	if (options.cursor) await handleCursorAction();
	else if (options.claudeCode) handleClaudeCodeAction();
	else if (options.openCode) handleOpenCodeAction();
	else if (options.manual) handleManualAction();
	else showAllOptions();
}
async function handleCursorAction() {
	console.log(chalk.bold.blue("🚀 Adding Better Auth MCP to Cursor..."));
	const platform = os$1.platform();
	let openCommand;
	switch (platform) {
		case "darwin":
			openCommand = "open";
			break;
		case "win32":
			openCommand = "start";
			break;
		case "linux":
			openCommand = "xdg-open";
			break;
		default: throw new Error(`Unsupported platform: ${platform}`);
	}
	const remoteConfig = { url: REMOTE_MCP_URL };
	const encodedRemote = base64.encode(new TextEncoder().encode(JSON.stringify(remoteConfig)));
	const remoteDeeplink = `cursor://anysphere.cursor-deeplink/mcp/install?name=${encodeURIComponent("better-auth")}&config=${encodedRemote}`;
	try {
		execSync(platform === "win32" ? `start "" "${remoteDeeplink}"` : `${openCommand} "${remoteDeeplink}"`, { stdio: "inherit" });
		console.log(chalk.green("\n✓ Better Auth MCP server installed!"));
	} catch {
		console.log(chalk.yellow("\n⚠ Could not automatically open Cursor for MCP installation."));
	}
	console.log(chalk.bold.white("\n✨ Next Steps:"));
	console.log(chalk.gray("• The MCP server will be added to your Cursor configuration"));
	console.log(chalk.gray("• You can now use Better Auth features directly in Cursor"));
	console.log(chalk.gray("• Try: \"Set up Better Auth with Google login\" or \"Help me debug my auth\""));
}
function handleClaudeCodeAction() {
	console.log(chalk.bold.blue("🤖 Adding Better Auth MCP to Claude Code..."));
	const command = `claude mcp add --transport http better-auth ${REMOTE_MCP_URL}`;
	try {
		execSync(command, { stdio: "inherit" });
		console.log(chalk.green("\n✓ Claude Code MCP configured!"));
	} catch {
		console.log(chalk.yellow("\n⚠ Could not automatically add to Claude Code. Please run this command manually:"));
		console.log(chalk.cyan(command));
	}
	console.log(chalk.bold.white("\n✨ Next Steps:"));
	console.log(chalk.gray("• The MCP server will be added to your Claude Code configuration"));
	console.log(chalk.gray("• You can now use Better Auth features directly in Claude Code"));
}
function handleOpenCodeAction() {
	console.log(chalk.bold.blue("🔧 Adding Better Auth MCP to Open Code..."));
	const openCodeConfig = {
		$schema: "https://opencode.ai/config.json",
		mcp: { "better-auth": {
			type: "remote",
			url: REMOTE_MCP_URL,
			enabled: true
		} }
	};
	const configPath = path$1.join(process.cwd(), "opencode.json");
	try {
		let existingConfig = {};
		if (fs$2.existsSync(configPath)) {
			const existingContent = fs$2.readFileSync(configPath, "utf8");
			existingConfig = JSON.parse(existingContent);
		}
		const mergedConfig = {
			...existingConfig,
			...openCodeConfig,
			mcp: {
				...existingConfig.mcp,
				...openCodeConfig.mcp
			}
		};
		fs$2.writeFileSync(configPath, JSON.stringify(mergedConfig, null, 2));
		console.log(chalk.green(`\n✓ Open Code configuration written to ${configPath}`));
		console.log(chalk.green("✓ Better Auth MCP server added successfully!"));
	} catch {
		console.log(chalk.yellow("\n⚠ Could not automatically write opencode.json. Please add this configuration manually:"));
		console.log(chalk.cyan(JSON.stringify(openCodeConfig, null, 2)));
	}
	console.log(chalk.bold.white("\n✨ Next Steps:"));
	console.log(chalk.gray("• Restart Open Code to load the new MCP server"));
	console.log(chalk.gray("• You can now use Better Auth features directly in Open Code"));
}
function handleManualAction() {
	console.log(chalk.bold.blue("📝 Better Auth MCP Configuration..."));
	const manualConfig = { "better-auth": { url: REMOTE_MCP_URL } };
	const configPath = path$1.join(process.cwd(), "mcp.json");
	try {
		let existingConfig = {};
		if (fs$2.existsSync(configPath)) {
			const existingContent = fs$2.readFileSync(configPath, "utf8");
			existingConfig = JSON.parse(existingContent);
		}
		const mergedConfig = {
			...existingConfig,
			...manualConfig
		};
		fs$2.writeFileSync(configPath, JSON.stringify(mergedConfig, null, 2));
		console.log(chalk.green(`\n✓ MCP configuration written to ${configPath}`));
		console.log(chalk.green("✓ Better Auth MCP server added successfully!"));
	} catch {
		console.log(chalk.yellow("\n⚠ Could not automatically write mcp.json. Please add this configuration manually:"));
		console.log(chalk.cyan(JSON.stringify(manualConfig, null, 2)));
	}
	console.log(chalk.bold.white("\n✨ Next Steps:"));
	console.log(chalk.gray("• Restart your MCP client to load the new server"));
	console.log(chalk.gray("• You can now use Better Auth features directly in your MCP client"));
}
function showAllOptions() {
	console.log(chalk.bold.blue("🔌 Better Auth MCP Server"));
	console.log(chalk.gray("Choose your MCP client to get started:"));
	console.log();
	console.log(chalk.bold.white("MCP Clients:"));
	console.log(chalk.cyan("  --cursor      ") + chalk.gray("Add to Cursor"));
	console.log(chalk.cyan("  --claude-code ") + chalk.gray("Add to Claude Code"));
	console.log(chalk.cyan("  --open-code   ") + chalk.gray("Add to Open Code"));
	console.log(chalk.cyan("  --manual      ") + chalk.gray("Manual configuration"));
	console.log();
	console.log(chalk.bold.white("Server:"));
	console.log(chalk.gray("  • ") + chalk.white("better-auth") + chalk.gray(" - Search documentation, code examples, setup assistance"));
	console.log();
}
const mcp = new Command("mcp").description("Add Better Auth MCP server to MCP Clients").option("--cursor", "Automatically open Cursor with the MCP configuration").option("--claude-code", "Show Claude Code MCP configuration command").option("--open-code", "Show Open Code MCP configuration").option("--manual", "Show manual MCP configuration for mcp.json").action(mcpAction);

//#endregion
//#region src/commands/migrate.ts
/** @internal */
async function migrateAction(opts) {
	const options = z$1.object({
		cwd: z$1.string(),
		config: z$1.string().optional(),
		y: z$1.boolean().optional(),
		yes: z$1.boolean().optional()
	}).parse(opts);
	const cwd = path.resolve(options.cwd);
	if (!existsSync(cwd)) {
		console.error(`The directory "${cwd}" does not exist.`);
		process.exit(1);
	}
	const config = await getConfig({
		cwd,
		configPath: options.config
	});
	if (!config) {
		console.error("No configuration file found. Add a `auth.ts` file to your project or pass the path to the configuration file using the `--config` flag.");
		return;
	}
	const db = await getAdapter(config);
	if (!db) {
		console.error("Invalid database configuration. Make sure you're not using adapters. Migrate command only works with built-in Kysely adapter.");
		process.exit(1);
	}
	if (db.id !== "kysely") {
		if (db.id === "prisma") {
			console.error("The migrate command only works with the built-in Kysely adapter. For Prisma, run `npx auth generate` to create the schema, then use Prisma's migrate or push to apply it.");
			try {
				await (await createTelemetry(config)).publish({
					type: "cli_migrate",
					payload: {
						outcome: "unsupported_adapter",
						adapter: "prisma",
						config: await getTelemetryAuthConfig(config)
					}
				});
			} catch {}
			process.exit(0);
		}
		if (db.id === "drizzle") {
			console.error("The migrate command only works with the built-in Kysely adapter. For Drizzle, run `npx auth generate` to create the schema, then use Drizzle's migrate or push to apply it.");
			try {
				await (await createTelemetry(config)).publish({
					type: "cli_migrate",
					payload: {
						outcome: "unsupported_adapter",
						adapter: "drizzle",
						config: await getTelemetryAuthConfig(config)
					}
				});
			} catch {}
			process.exit(0);
		}
		console.error("Migrate command isn't supported for this adapter.");
		try {
			await (await createTelemetry(config)).publish({
				type: "cli_migrate",
				payload: {
					outcome: "unsupported_adapter",
					adapter: db.id,
					config: await getTelemetryAuthConfig(config)
				}
			});
		} catch {}
		process.exit(1);
	}
	const spinner = yoctoSpinner({ text: "preparing migration..." }).start();
	const { toBeAdded, toBeCreated, runMigrations } = await getMigrations(config);
	if (!toBeAdded.length && !toBeCreated.length) {
		spinner.stop();
		console.log("🚀 No migrations needed.");
		try {
			await (await createTelemetry(config)).publish({
				type: "cli_migrate",
				payload: {
					outcome: "no_changes",
					config: await getTelemetryAuthConfig(config)
				}
			});
		} catch {}
		process.exit(0);
	}
	spinner.stop();
	console.log(`🔑 The migration will affect the following:`);
	for (const table of [...toBeCreated, ...toBeAdded]) console.log("->", chalk.magenta(Object.keys(table.fields).join(", ")), chalk.white("fields on"), chalk.yellow(`${table.table}`), chalk.white("table."));
	if (options.y) {
		console.warn("WARNING: --y is deprecated. Consider -y or --yes");
		options.yes = true;
	}
	let migrate = options.yes;
	if (!migrate) migrate = (await prompts({
		type: "confirm",
		name: "migrate",
		message: "Are you sure you want to run these migrations?",
		initial: false
	})).migrate;
	if (!migrate) {
		console.log("Migration cancelled.");
		try {
			await (await createTelemetry(config)).publish({
				type: "cli_migrate",
				payload: {
					outcome: "aborted",
					config: await getTelemetryAuthConfig(config)
				}
			});
		} catch {}
		process.exit(0);
	}
	spinner?.start("migrating...");
	await runMigrations();
	spinner.stop();
	console.log("🚀 migration was completed successfully!");
	try {
		await (await createTelemetry(config)).publish({
			type: "cli_migrate",
			payload: {
				outcome: "migrated",
				config: await getTelemetryAuthConfig(config)
			}
		});
	} catch {}
	process.exit(0);
}
const migrate = new Command("migrate").option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).option("--config <config>", "the path to the configuration file. defaults to the first configuration file found.").option("-y, --yes", "automatically accept and run migrations without prompting", false).option("--y", "(deprecated) same as --yes", false).action(migrateAction);

//#endregion
//#region src/commands/secret.ts
const generateSecret = new Command("secret").action(() => {
	const secret = generateSecretHash();
	console.log(`\nAdd the following to your .env file: 
${chalk.gray("# Auth Secret") + chalk.green(`\nBETTER_AUTH_SECRET=${secret}`)}`);
});
const generateSecretHash = () => {
	return Crypto.randomBytes(32).toString("hex");
};

//#endregion
//#region src/index.ts
process.on("SIGINT", () => process.exit(0));
process.on("SIGTERM", () => process.exit(0));
let cliVersion = "1.1.2";
async function main() {
	const program = new Command("better-auth");
	let packageInfo = {};
	try {
		packageInfo = await getPackageInfo();
		cliVersion = packageInfo.version || "1.1.2";
	} catch {}
	program.addCommand(init).addCommand(migrate).addCommand(generate).addCommand(generateSecret).addCommand(info).addCommand(login).addCommand(logout).addCommand(mcp).version(cliVersion).description("Better Auth CLI").action(() => program.help());
	program.parse();
}
main().catch((error) => {
	console.error("Error running Better Auth CLI:", error);
	process.exit(1);
});

//#endregion
export { cliVersion };
